<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Introduction to R for Biologists: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="favicons/mb/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicons/mb/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicons/mb/favicon-16x16.png">
<link rel="manifest" href="favicons/mb/site.webmanifest">
<link rel="mask-icon" href="favicons/mb/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav mb"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="mb-logo" alt="Lesson Description" src="assets/images/mb-logo.svg"><span class="badge text-bg-info">
          <abbr title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.">
            <a href="https://docs.carpentries.org/resources/curriculum/lesson-life-cycle.html" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-circle" style="border-radius: 5px"></i>
              Beta
            </a>
            <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>


      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav mb" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="assets/images/mb-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Introduction to R for Biologists
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Introduction to R for Biologists
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to R for Biologists
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress mb">
    <div class="progress-bar mb" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction to R</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="explore-data.html">2. Exploring the Data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="format-data.html">3. Formatting the Data</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="ggplot2-intro.html">4. Introduction to ggplot2</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="ggplot2-customisation.html">5. Extra ggplot2 Customisation</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="wrap-up.html">6. Wrapping Up</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Introduction to R</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p><strong>Authors:</strong> Maria Doyle, Jessica Chung, Vicky Perreau,
Kim-Anh Lê Cao, Saritha Kodikara, Eva Hamrud.</p>
<p>These wonderful people wrote this content.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why might you prefer to run code in a script rather than directly
from the console?</li>
<li>What is the assignment operator and how can you use it to store
objects?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Introduce the R programming language and RStudio</li>
<li>Explain functions, objects, and packages</li>
<li>Demonstrate how to view help pages and documentation</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="r-for-biologists-course">R for Biologists course<a class="anchor" aria-label="anchor" href="#r-for-biologists-course"></a>
</h2>
<hr class="half-width">
<p>R takes time to learn, like a spoken language. No one can expect to
be an R expert after learning R for a few hours. This course has been
designed to introduce biologists to R, showing some basics, and also
some powerful things R can do (things that would be more difficult to do
with Excel). The aim is to give beginners the confidence to continue
learning R, so the focus here is on tidyverse and visualisation of
biological data, as we believe this is a productive and engaging way to
start learning R. After this short introduction you could use <a href="https://bookdown.org/ansellbr/WEHI_tidyR_course_book" class="external-link">this
book</a> to dive a bit deeper.</p>
<p>Most R programmers do not remember all the command lines we share in
this document. R is a language that is continuously evolving. They use
Google extensively to use many new tricks. Do not hesitate to do the
same!</p>
</section><section><h2 class="section-heading" id="intro-to-r-and-rstudio">Intro to R and RStudio<a class="anchor" aria-label="anchor" href="#intro-to-r-and-rstudio"></a>
</h2>
<hr class="half-width">
<p>RStudio is an interface that makes it easier to use R. There are four
panels in RStudio. The screenshot below shows an <a href="https://twitter.com/RLadiesNCL/status/1138812826917724160" class="external-link">analogy
linking the different RStudio panels to cooking</a>.</p>
<figure><img src="fig/rstudio_cooking.jpg" class="figure mx-auto d-block"></figure></section><section><h2 class="section-heading" id="r-script-vs-console">R script vs console<a class="anchor" aria-label="anchor" href="#r-script-vs-console"></a>
</h2>
<hr class="half-width">
<p>There are two ways to work in RStudio in the console or in a
script.</p>
<p>Let’s start by running a command in the console.</p>
<div id="your-turn-1.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.1" class="callout-inner">
<h3 class="callout-title">Your turn 1.1</h3>
<div class="callout-content">
<p>Run the command below in the console.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">+</span> <span class="fl">1</span></span></code></pre>
</div>
<p>Once you’ve typed in the command into your console, just press
<kbd>enter</kbd>. The output should be printed into the console.</p>
</div>
</div>
</div>
<p>Alternatively, we can use an R script. An R script allows us to have
a record (recipe) of what we have done, whilst commands we type into the
console are not saved. Keeping a record speeds up our analysis because
we can re-use code, its also helpful to remember what we have done
before!</p>
<div id="your-turn-1.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.2" class="callout-inner">
<h3 class="callout-title">Your turn 1.2</h3>
<div class="callout-content">
<p>Create a script from the top menu in RStudio:<br><code>File &gt; New File &gt; R Script</code>, then type the command
below in your new script.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">2</span> <span class="op">+</span> <span class="fl">2</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>To run a command in a script, we place the cursor on the line you
want to run, and then either:</p>
<ul>
<li>Click on the <code>run</code> button on top of the panel</li>
<li>Use <kbd>Ctrl</kbd> + <kbd>Enter</kbd> (Windows/Linux) or
<kbd>Cmd</kbd> + <kbd>Enter</kbd> (MacOS).</li>
</ul>
<p>You can also highlight multiple lines at once and run them at once -
have a go!</p>
<div id="recommended-practice-use-scripts" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="recommended-practice-use-scripts" class="callout-inner">
<h3 class="callout-title">Recommended Practice: Use Scripts</h3>
<div class="callout-content">
<p>The console is great for quick tests or running single lines of code.
However, for more complex analyses, using scripts is best practice.
Scripts help keep your work organized and reproducible. We’ll use
scripts for the rest of this workshop.</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="commenting">Commenting<a class="anchor" aria-label="anchor" href="#commenting"></a>
</h2>
<hr class="half-width">
<p>Comments are notes to ourself or others about the commands in the
script. They are useful also when you share code with others. Comments
start with a <code>#</code> which tells R not to run them as
commands.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># testing R</span></span>
<span><span class="fl">2</span> <span class="op">+</span> <span class="fl">2</span></span></code></pre>
</div>
<p>Keeping an accurate record of how you have manipulated your data is
important for <a href="https://ropensci-archive.github.io/reproducibility-guide/" class="external-link">reproducible
research</a>. Writing detailed comments and documenting your work are
useful reminders to your future self (and anyone else reading your
scripts) on what your code does.</p>
<p>Commenting code is good practice, why not try commenting on the code
you write in this session to get into the habit, it will also make your
R script more informative when you come back to it in the future.</p>
<hr></section><section><h2 class="section-heading" id="working-directory">Working directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h2>
<hr class="half-width">
<p>Opening an RStudio session launches it from a specific location. This
is the ‘working directory’.</p>
<div id="understanding-the-working-directory" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="understanding-the-working-directory" class="callout-inner">
<h3 class="callout-title">Understanding the Working Directory</h3>
<div class="callout-content">
<p>The working directory is the folder where R reads and saves files on
your computer. When working in R, you’ll often read data files and write
outputs like analysis results or plots. Knowing where your working
directory is set helps ensure R finds your files and saves outputs in
the right place.</p>
</div>
</div>
</div>
<p>You can find out where your current working directory is set to by
using two different approaches:</p>
<ul>
<li>Run the command <code>getwd()</code>. This will print out the path
to your working directory in the console. It will be in this format:
<code>/path/to/working/directory</code> (Mac) or
<code>C:\path\to\working\directory</code> (Windows), or</li>
<li>In the bottom-right panel, click the blue cog icon on the menu at
the top, then click <code>Go To Working Directory</code>. This will show
you the location and files in your working directory in the files
window.</li>
</ul>
<div id="your-turn-1.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.3" class="callout-inner">
<h3 class="callout-title">Your turn 1.3</h3>
<div class="callout-content">
<p>Where is you working directory set to at the moment? Is this a useful
place to have it set?</p>
</div>
</div>
</div>
<p>By default the working directory is often your home directory. To
keep data and scripts organised its good practice to set your working
directory as a specific folder.</p>
<div id="your-turn-1.4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.4" class="callout-inner">
<h3 class="callout-title">Your turn 1.4</h3>
<div class="callout-content">
<p>Create a folder for this course somewhere on your computer. Name the
folder something meaningful, for example, <code>intro_r_course</code> or
<code>Introduction_to_R</code>. Then, to set this folder as your working
directory, you can do this in multiple ways, e.g.:</p>
<ul>
<li>Click in the menu at the top on
<code>Session &gt; Set Working Directory &gt; Choose directory</code>
and choose your folder, or</li>
<li>In the bottom-right panel, navigate to the folder that you want to
be your working directory. You can also do this by clicking on the three
dots icon on the top-right of the panel. Then once you’re in a suitable
directory, in the menu bar of the in the bottom-right panel, navigate to
the blue cog icon and click <code>Set As Working Directory</code>
</li>
</ul>
<figure><img src="fig/three_dots.png" class="figure mx-auto d-block"></figure><figure><img src="fig/working_directory.png" class="figure mx-auto d-block"></figure><p>You will see that once you have set your working directory, the files
inside your new folder will appear in the ‘Files’ window on RStudio.</p>
</div>
</div>
</div>
<div id="your-turn-1.5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.5" class="callout-inner">
<h3 class="callout-title">Your turn 1.5</h3>
<div class="callout-content">
<p>Save the script you created in the previous section as
<code>intro.R</code> in this directory. You can do this by clicking on
<code>File &gt; Save</code> and the default location should be the
current working directory (e.g. <code>intro_r_course</code>).</p>
</div>
</div>
</div>
<div id="multiple-ways-to-achieve-the-same-goal-in-r" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="multiple-ways-to-achieve-the-same-goal-in-r" class="callout-inner">
<h3 class="callout-title">Multiple Ways to Achieve the Same Goal in
R</h3>
<div class="callout-content">
<p>You might have noticed by now that in R, there are often several ways
to accomplish the same task. You might find one method more intuitive or
easier to use than others — and that’s okay! Experiment, explore, and
choose the approach that works best for you.</p>
</div>
</div>
</div>
<p>You might have noticed that when you set your working directory in
the previous step, a line appeared in your console saying something like
<code>setwd("~/Desktop/intro_r_course")</code>. As well as the
point-and-click methods described above, you can also set your working
directory using the <code>setwd()</code> command in the console or in a
script.</p>
<div id="your-turn-1.6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-1.6" class="callout-inner">
<h3 class="callout-title">Your turn 1.6</h3>
<div class="callout-content">
<p>What might be an advantage of using the command line option
(i.e. <code>setwd()</code>) over point-and-click methods to set your
working directory?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<p>There is no easy way to record what you point and click on (unless
you write it all down!). Putting a command at the top of the script
means you are less likely to forget where you have your working
directory, and when you come back to it another day you can quickly
re-run it.</p>
</div>
</div>
</div>
</div>
<div id="your-turn-1.6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.6" class="callout-inner">
<h3 class="callout-title">Your turn 1.6<em> (continued)</em>
</h3>
<div class="callout-content">
<p>Add a line at the top of your newly created script
<code>intro.R</code> so that the working directory is set to your newly
made folder (e.g. <code>intro_r_course</code>).</p>
</div>
</div>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> RStudio Projects </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<p>You can also use RStudio projects as described <a href="https://r4ds.had.co.nz/workflow-projects.html" class="external-link">here</a> to
automatically keep track of and set the working directory.</p>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a>
</h2>
<hr class="half-width">
<p>In mathematics, a function defines a relation between inputs and
output. In R (and other coding languages) it is the same. A function
(also called a command) takes inputs called <em>arguments</em> inside
parentheses, and output some results.</p>
<p>We have actually already used two functions in this workshop -
<code>getwd()</code> and <code>setwd()</code>. <code>getwd()</code> does
not take an input, but outputs your working directory.
<code>setwd()</code> takes a path as its input, and sets it as your
working directory.</p>
<p>Let’s take a look at some more functions below.</p>
<div id="your-turn-1.7" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.7" class="callout-inner">
<h3 class="callout-title">Your turn 1.7</h3>
<div class="callout-content">
<p>Compare these two outputs. In the second line we use the function
<code>sum()</code>.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">2</span> <span class="op">+</span> <span class="fl">2</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-1.8" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.8" class="callout-inner">
<h3 class="callout-title">Your turn 1.8</h3>
<div class="callout-content">
<p>Try using the below function with different inputs, what does it
do?</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sqrt</span><span class="op">(</span><span class="fl">9</span><span class="op">)</span></span>
<span><span class="fu">sqrt</span><span class="op">(</span><span class="fl">81</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="tab-completion" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tab-completion" class="callout-inner">
<h3 class="callout-title">Tab completion</h3>
<div class="callout-content">
<p>A very useful feature is Tab completion. You can start typing and use
<kbd>Tab</kbd> to autocomplete code, for example, a function name.</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="objects">Objects<a class="anchor" aria-label="anchor" href="#objects"></a>
</h2>
<hr class="half-width">
<p>It is useful to store data or results so that we can use them later
on for other parts of the analysis. To do this, we can store data as
<strong>objects</strong>. We can use the assignment operator
<code>&lt;-</code>, where the name of the object (which are called
<strong>variables</strong>) is on the left side of the arrow, and the
data you want to store is on the right side.</p>
<p>For example, the below code assigns the number <code>5</code> to the
object <code>x</code> using the <code>&lt;-</code> operator. You can
print out what the <code>x</code> object is by just typing it into the
console or running it in your script.</p>
<div id="your-turn-1.9" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.9" class="callout-inner">
<h3 class="callout-title">Your turn 1.9</h3>
<div class="callout-content">
<p>Play around and create some objects. Then print out the objects using
their names.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">x</span></span>
<span></span>
<span><span class="va">result_1</span> <span class="op">&lt;-</span> <span class="fl">2</span> <span class="op">+</span> <span class="fl">2</span></span>
<span><span class="va">result_1</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="assignment-operator-shortcut" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="assignment-operator-shortcut" class="callout-inner">
<h3 class="callout-title">Assignment operator shortcut</h3>
<div class="callout-content">
<p>In RStudio, typing <kbd>Alt</kbd> + <kbd>-</kbd> (holding down
<kbd>Alt</kbd> at the same time as the <kbd>-</kbd> key) will write
<code>&lt;-</code> in a single keystroke in Windows, while typing &gt;
<kbd>Option</kbd> + <kbd>-</kbd> (holding down <kbd>Option</kbd> at the
same time as the <kbd>-</kbd> key) does the same in a Mac.</p>
</div>
</div>
</div>
<p>Once you have assigned objects, you can perform manipulations on them
using functions.</p>
<div id="your-turn-1.10" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.10" class="callout-inner">
<h3 class="callout-title">Your turn 1.10</h3>
<div class="callout-content">
<p>Compare the two outputs.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span><span class="fu">sum</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Remember, if you use the same object name multiple times, R will
overwrite the previous object you had created.</p>
<div id="your-turn-1.11" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-1.11" class="callout-inner">
<h3 class="callout-title">Your turn 1.11</h3>
<div class="callout-content">
<p>What is the value of <code>x</code> after running this code?</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">10</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<p><code>x</code> is 10. The previous value of 5 has been
overwritten.</p>
</div>
</div>
</div>
</div>
<div id="your-turn-1.12" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-1.12" class="callout-inner">
<h3 class="callout-title">Your turn 1.12</h3>
<div class="callout-content">
<p>Can you write some code to calculate the sum of the square root of 9
and the square root of 16?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sum</span><span class="op">(</span><span class="fu">sqrt</span><span class="op">(</span><span class="fl">9</span><span class="op">)</span>, <span class="fu">sqrt</span><span class="op">(</span><span class="fl">16</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 7</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="recommended-practice-name-your-objects-with-care" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="recommended-practice-name-your-objects-with-care" class="callout-inner">
<h3 class="callout-title">Recommended Practice: Name your objects with
care</h3>
<div class="callout-content">
<p>Use clear and descriptive names for your objects, such as
<code>data_raw</code> and <code>data_normalised</code>, instead of vague
names like <code>data1</code> and <code>data2</code>. This makes it
easier to track different steps in your analysis.</p>
</div>
</div>
</div>
<p>So far we have looked at objects which are numbers. However objects
can also be made of characters, these are called <em>strings</em>. To
create a string you need to use quotation marks <code>""</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_string</span> <span class="op">&lt;-</span> <span class="st">"Hello!"</span></span>
<span><span class="va">my_string</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Hello!"</code></pre>
</div>
<p>There are a whole host of different objects you can make in R, too
many to cover in this session! Later on when we do some data wrangling
we will work with objects which are <em>dataframes</em> (i.e. tables)
and <em>vectors</em> (a series of numbers and/or strings). Let’s make a
simple vector now to get familiar. To make a vector you need to use the
command <code>c()</code>.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_vector</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">my_vector</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_new_vector</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Hello"</span>, <span class="st">"World"</span><span class="op">)</span></span>
<span><span class="va">my_new_vector</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Hello" "World"</code></pre>
</div>
<div id="your-turn-1.13" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-1.13" class="callout-inner">
<h3 class="callout-title">Your turn 1.13</h3>
<div class="callout-content">
<p>Try making an object and setting it as <code>1:5</code>, what does
this object look like?</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">5</span></span>
<span><span class="va">x</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1 2 3 4 5</code></pre>
</div>
<p><code>1:5</code> creates a vector with a sequence of numbers from 1
to 5.</p>
</div>
</div>
</div>
</div>
<p>Once you have a vector, you can subset it. We will cover this further
when we do some data wrangling but lets try a simple example here.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_vector</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span><span class="co"># extract the first element from the vector</span></span>
<span><span class="va">my_vector</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "A"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the last element from the vector</span></span>
<span><span class="va">my_vector</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "C"</code></pre>
</div>
<div id="your-turn-1.14" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-1.14" class="callout-inner">
<h3 class="callout-title">Your turn 1.14</h3>
<div class="callout-content">
<p>Create a vector from 1 to 10 and print the 9th element of the
vector.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_vector</span> <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span></span>
<span><span class="va">my_vector</span><span class="op">[</span><span class="fl">9</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 9</code></pre>
</div>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="packages">Packages<a class="anchor" aria-label="anchor" href="#packages"></a>
</h2>
<hr class="half-width">
<p>We have seen that functions are really useful tools which can be used
to manipulate data. Although some basic functions, like
<code>sum()</code> and <code>setwd()</code> are available by default
when you install R, some more exciting functions are not. There are
thousands of R functions available for you to use, and functions are
organised into groups called <em>packages</em> or <em>libraries</em>. An
R package contains a collection of functions (usually that perform
related tasks), as well as documentation to explain how to use the
functions. Packages are made by R developers who wish to share their
methods with others.</p>
<p>Once we have identified a package we want to use, we can install and
load it so we can use it. Here we will use the <code>tidyverse</code>
package which includes lots of useful functions for data managing, we
will use the package later in this session.</p>
<p>If it’s not already installed on your computer, you can use the
<code>install.packages</code> <strong>function</strong> to install a
<strong>package</strong>. A package is a collection of functions along
with documentation, code, tests and example data.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<hr></section><section><h2 class="section-heading" id="packages-in-the-cran-or-bioconductor">Packages in the CRAN or Bioconductor<a class="anchor" aria-label="anchor" href="#packages-in-the-cran-or-bioconductor"></a>
</h2>
<hr class="half-width">
<p>Packages are hosted in different locations. Packages hosted on <a href="https://cran.r-project.org/" class="external-link">CRAN</a> (stands for Comprehensive R
Archive Network) are often generic package for all sorts of data and
analysis. <a href="https://bioconductor.org/" class="external-link">Bioconductor</a> is an
ecosystem that hosts packages specifically dedicated to biological
data.</p>
<p>The installation of packages from Bioconductor is a bit different,
e.g to install the <code>mixOmics</code> package we type:</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You don't need to run this codeblock for this workshop</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw">require</span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu">install.packages</span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"mixOmics"</span><span class="op">)</span></span></code></pre>
</div>
<p>You don’t need to remember this command line, as it is featured in
the Bioconductor package page (see <a href="https://bioconductor.org/packages/release/bioc/html/mixOmics.html" class="external-link">here</a>
for example).</p>
<p>One advantage of Bioconductor packages is that they are well
documented, updated and maintained every six months.</p>
<hr></section><section><h2 class="section-heading" id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<hr class="half-width">
<p>As described above, every R package includes documentation to explain
how to use functions. For example, to find out what a function in R
does, type a <code>?</code> before the name and help information will
appear in the Help panel on the right in RStudio.</p>
<div id="your-turn-1.15" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.15" class="callout-inner">
<h3 class="callout-title">Your turn 1.15</h3>
<div class="callout-content">
<p>Find out what the <code>sum()</code> command does.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">sum</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>What is really important is to scroll down the examples to understand
how the function can be used in practice. You can use this command line
to run the examples:</p>
<div id="your-turn-1.16" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.16" class="callout-inner">
<h3 class="callout-title">Your turn 1.16</h3>
<div class="callout-content">
<p>Run some examples of the <code>sum()</code> command.</p>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">example</span><span class="op">(</span><span class="va">sum</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Packages also come with more comprehensive documentation called
vignettes. These are really helpful to get you started with the package
and identify which functions you might want to use.</p>
<div id="your-turn-1.17" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-1.17" class="callout-inner">
<h3 class="callout-title">Your turn 1.17</h3>
<div class="callout-content">
<p>Have a look at the <code>tidyverse</code> package vignette.</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">browseVignettes</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="common-r-errors">Common R errors<a class="anchor" aria-label="anchor" href="#common-r-errors"></a>
</h2>
<hr class="half-width">
<p>R error messages are common and often cryptic. You most likely will
encounter at least one error message during this tutorial. Some common
reasons for errors are:</p>
<ul>
<li>Case sensitivity. In R, as in other programming languages, case
sensitivity is important. <code>?install.packages</code> is different to
<code>?Install.packages</code>.</li>
<li>Missing commas</li>
<li>Mismatched parentheses or brackets or unclosed parentheses, brackets
or apostrophes</li>
<li>Not quoting file paths (<code>""</code>)</li>
<li>When a command line is unfinished, the “+” in the console will
indicate it is awaiting further instructions. Press <kbd>ESC</kbd> to
cancel the command.</li>
</ul>
<p>To see examples of some R error messages with explanations see <a href="https://github.com/noamross/zero-dependency-problems/issues/7" class="external-link">here</a></p>
<div id="more-information-for-when-you-get-stuck" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="more-information-for-when-you-get-stuck" class="callout-inner">
<h3 class="callout-title">More information for when you get stuck</h3>
<div class="callout-content">
<p>As well as using package vignettes and documentation, Google and <a href="https://stackoverflow.com/" class="external-link">Stack Overflow</a> are also useful
resources for getting help.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>Use scripts for analyses over typing commands in the console. This
allows you to keep an accurate record of what you did, which is
important for reproducible research</li>
<li>You can store data as objects using the assignment operator
<code>&lt;-</code>
</li>
<li>Installing packages from CRAN is done with the
<code>install.packages()</code> function</li>
<li>You can view the help page of a function by typing a <code>?</code>
before the name (e.g. <code>?sum</code>)</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-explore-data"><p>Content from <a href="explore-data.html">Exploring the Data</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/explore-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I load data from CSV or TSV files into R?</li>
<li>What are some functions in R that can be used to examine the
data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Load in our RNA-seq data files into R</li>
<li>Try using some functions to explore data frames (tables) in R</li>
<li>Learn how to subset parts of a data frame</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="getting-started-with-the-data">Getting started with the data<a class="anchor" aria-label="anchor" href="#getting-started-with-the-data"></a>
</h2>
<hr class="half-width">
<p>In this tutorial, we will learn some R through creating plots to
visualise data from an RNA-seq experiment.</p>
<p>The <a href="https://www.nature.com/articles/s41598-019-43935-8" class="external-link">GREIN
platform</a> (GEO RNA-seq Experiments Interactive Navigator) provides
&gt;6,500 published datasets from GEO that have been uniformly
processed. It is available at <a href="http://www.ilincs.org/apps/grein/" class="external-link uri">http://www.ilincs.org/apps/grein/</a>. You can search for a
dataset of interest using the GEO code. GREIN provide QC metrics for the
RNA-seq datasets and both raw and normalized counts. We will use the
normalized counts here. These are the counts of reads for each gene for
each sample normalized for differences in sequencing depth and
composition bias. Generally, the higher the number of counts the more
the gene is expressed.</p>
</section><section><h2 class="section-heading" id="rna-seq-dataset-from-fu-et-al-">RNA-seq dataset from Fu et al.<a class="anchor" aria-label="anchor" href="#rna-seq-dataset-from-fu-et-al-"></a>
</h2>
<hr class="half-width">
<p>Here we will create some plots using RNA-seq data from the paper by
<a href="https://www.ncbi.nlm.nih.gov/pubmed/25730472" class="external-link">Fu et
al. 2015</a> (GEO accession number GSE60450). Mice, like all mammals,
have mammary glands which produce milk to nourish their young. The
authors of this study were interested in examining how the mammary
epithelium (which line the mammary glands) expands and develops during
pregnancy and lactation.</p>
<p>This study examined expression in basal and luminal cells from mice
at different stages (virgin, pregnant and lactating). Basal cells are
mammary stem cells and luminal cells secrete milk. There are 2 samples
per group and 6 groups, so 12 samples in total.</p>
<figure><img src="fig/mouse_exp.png" class="figure mx-auto d-block"></figure><hr></section><section><h2 class="section-heading" id="tidyverse">Tidyverse<a class="anchor" aria-label="anchor" href="#tidyverse"></a>
</h2>
<hr class="half-width">
<figure><img src="fig/tidyverse.png" style="width:100.0%" class="figure mx-auto d-block"></figure><p>The <code>tidyverse</code> package that we installed previously is a
collection of R packages that includes the extremely widely used
<strong><code>ggplot2</code></strong> package.</p>
<p>The <a href="https://www.tidyverse.org/" class="external-link">tidyverse</a> makes data
science faster, easier and more fun.</p>
<figure><img src="fig/r_tidyverse.png" style="width:100.0%" class="figure mx-auto d-block"></figure><p>Tidyverse is built on the principle of organizing data in a <a href="https://www.openscapes.org/blog/2020/10/12/tidy-data/" class="external-link">tidy
format</a>.</p>
<hr></section><section><h2 class="section-heading" id="data-files">Data files<a class="anchor" aria-label="anchor" href="#data-files"></a>
</h2>
<hr class="half-width">
<div id="your-turn-2.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.1" class="callout-inner">
<h3 class="callout-title">Your turn 2.1</h3>
<div class="callout-content">
<p>If you haven’t already downloaded the data.zip file for this
workshop, you can click <a href="data/data.zip">here</a> to download it
now.</p>
<p>Unzip the file and store the extracted <code>data</code> folder in
your working directory.</p>
</div>
</div>
</div>
<div id="make-sure-the-data-is-in-the-correct-directory" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="make-sure-the-data-is-in-the-correct-directory" class="callout-inner">
<h3 class="callout-title">Make sure the data is in the correct
directory</h3>
<div class="callout-content">
<p>Inside your current working directory directory
(e.g. <code>intro_r_course</code>), there should be a directory called
<code>data</code> and inside that directory should be two CSV files.</p>
<figure><img src="fig/working_dir_files.png" style="width:80.0%" class="figure mx-auto d-block"></figure><p>In the next section, you will load these files into R using the file
path (i.e. where they are located in the filesystem), so the code in
this tutorial expects the files to be in the above location. If you have
the files in a different location, you also have the option of changing
the commands in the following sections to match where the files are
located on your computer.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h2>
<hr class="half-width">
<p>We use <code>library()</code> to load in the packages that we need.
As described in the cooking analogy in the first screenshot,
<code>install.packages()</code> is like buying a saucepan,
<code>library()</code> is taking it out of the cupboard to use it.</p>
<div id="your-turn-2.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.2" class="callout-inner">
<h3 class="callout-title">Your turn 2.2</h3>
<div class="callout-content">
<p>Load in the tidyverse package using the <code>library()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>The files we will use are CSV comma-separated, so we will use the
<code>read_csv()</code> function from the tidyverse. There is also a
<code>read_tsv()</code> function for tab-separated values.</p>
<p>We will use the counts file called
<code>GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv</code> that’s
in a folder called <code>data</code> i.e. the path to the file should be
<code>data/GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv</code>.</p>
<p>We can read the counts file into R with the command below. We’ll
store the contents of the counts file in an <strong>object</strong>
called <code>counts</code>. This stores the file contents in R’s memory
making it easier to use.</p>
<div id="your-turn-2.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.3" class="callout-inner">
<h3 class="callout-title">Your turn 2.3</h3>
<div class="callout-content">
<p>Load the count data into R. We will store the contents of the counts
file in an <strong>object</strong> called <code>counts</code>. Note that
we need to put quotes (““) around file paths.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in counts file</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>New names:
Rows: 23735 Columns: 14
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(2): ...1, gene_symbol dbl (12): GSM1480291, GSM1480292, GSM1480293,
GSM1480294, GSM1480295, GSM148...
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `` -&gt; `...1`</code></pre>
</div>
<p>Note: In R, structured tables like these are called <strong>data
frames</strong>.</p>
</div>
</div>
</div>
<div id="tab-completion-for-file-paths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tab-completion-for-file-paths" class="callout-inner">
<h3 class="callout-title">Tab completion for file paths</h3>
<div class="callout-content">
<p>We mentioned tab completion in the previous section, but tab
completion can also complete file paths. This means you don’t have to
type out the long filenames in the codeblock above, but instead, begin
to type out a few characters, then press <kbd>tab</kbd> and see the
autocompletion options.</p>
<figure><img src="fig/file_path_autocompletion.png" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<p>No need to be overwhelmed by the outputs! It contains information
regarding “column specification” (telling us that there is a missing
column name in the header and it has been filled with the name “…1”,
which is how read_csv handles missing column names by default). We will
fix this later. It also tells us what data types <code>read_csv</code>
is detecting in each column. Columns with text characters have been
detected (<code>col_character</code>) and also columns with numbers
(<code>col_double</code>). We won’t get into the details of R data types
in this tutorial but they are important to know when you get more
proficient in R. You can read more about them in the <a href="https://r4ds.had.co.nz/vectors.html#important-types-of-atomic-vector" class="external-link">R
for Data Science book</a>.</p>
<div id="your-turn-2.4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.4" class="callout-inner">
<h3 class="callout-title">Your turn 2.4</h3>
<div class="callout-content">
<p>Load the sample information data into R. We will store the contents
of this file in an object called <code>sampleinfo</code>.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in metadata</span></span>
<span><span class="va">sampleinfo</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/GSE60450_filtered_metadata.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>New names:
Rows: 12 Columns: 4
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(4): ...1, characteristics, immunophenotype, developmental stage
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `` -&gt; `...1`</code></pre>
</div>
</div>
</div>
</div>
<p>It is very common when looking at biological data that you have two
types of data. One is the actual data (in this case, our
<code>counts</code> object, which has the expression values of different
genes in each sample). The other is <strong>metadata</strong>
i.e. information about our samples (in this case, our
<code>sampleinfo</code> object includes information about whether
samples are from basal or luminal cells and whether the cells were from
mice which are virgin/pregnant/lactating, etc.)</p>
<p><strong>What data have we imported into R?</strong><br>
To summarise, we have imported two data frames (i.e. tables) into R:</p>
<ul>
<li>
<code>counts</code> object is our gene expression data</li>
<li>
<code>sampleinfo</code> object is our sample metadata</li>
</ul>
<div id="your-turn-2.5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.5" class="callout-inner">
<h3 class="callout-title">Your turn 2.5</h3>
<div class="callout-content">
<p>Let’s get used to making some mistakes in R, so we know what errors
look like and how to handle them.</p>
<ol style="list-style-type: decimal">
<li><p>Test what happens if you type
<code>Library(tidyverse)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type
<code>library(tidyverse</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_tsv("data/GSE60450_filtered_metadata.csv")</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_csv("data/GSE60450_filtered_metadata.csv)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_csv("GSE60450_filtered_metadata.csv")</code><br>
What is wrong and how would you fix it?</p></li>
</ol>
<p>Don’t forget you can press <kbd>ESC</kbd> to escape the current
command and start a new prompt.</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="getting-to-know-the-data">Getting to know the data<a class="anchor" aria-label="anchor" href="#getting-to-know-the-data"></a>
</h2>
<hr class="half-width">
<p>When assigning a value to an object, R does not print the value. We
do not see what is in <code>counts</code> or <code>sampleinfo</code>.
But there are ways we can look at the data.</p>
<div id="your-turn-2.6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.6" class="callout-inner">
<h3 class="callout-title">Your turn 2.6</h3>
<div class="callout-content">
<p>Click on the <code>sampleinfo</code> object in your global
environment panel on the right-hand-side of RStudio. This will open a
new tab.</p>
<p>This is the equivalent of using the <code>View()</code> function.
e.g.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-2.7" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.7" class="callout-inner">
<h3 class="callout-title">Your turn 2.7</h3>
<div class="callout-content">
<p>Type the name of the object and this will print the first few lines
and some information, such as number of rows. Note that this is similar
to how we looked at the value of objects we assigned in the previous
section.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 12 × 4
   ...1       characteristics              immunophenotype `developmental stage`
   &lt;chr&gt;      &lt;chr&gt;                        &lt;chr&gt;           &lt;chr&gt;
 1 GSM1480291 mammary gland, luminal cell… luminal cell p… virgin
 2 GSM1480292 mammary gland, luminal cell… luminal cell p… virgin
 3 GSM1480293 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 4 GSM1480294 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 5 GSM1480295 mammary gland, luminal cell… luminal cell p… 2 day lactation
 6 GSM1480296 mammary gland, luminal cell… luminal cell p… 2 day lactation
 7 GSM1480297 mammary gland, basal cells,… basal cell pop… virgin
 8 GSM1480298 mammary gland, basal cells,… basal cell pop… virgin
 9 GSM1480299 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
10 GSM1480300 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
11 GSM1480301 mammary gland, basal cells,… basal cell pop… 2 day lactation
12 GSM1480302 mammary gland, basal cells,… basal cell pop… 2 day lactation      </code></pre>
</div>
</div>
</div>
</div>
<p>We can also take a look the first few lines with <code>head()</code>.
This shows us the first 6 lines.</p>
<div id="your-turn-2.8" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.8" class="callout-inner">
<h3 class="callout-title">Your turn 2.8</h3>
<div class="callout-content">
<p>Use <code>head()</code> to look at the first few lines of
<code>counts</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 14
  ...1        gene_symbol GSM1480291 GSM1480292 GSM1480293 GSM1480294 GSM1480295
  &lt;chr&gt;       &lt;chr&gt;            &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 ENSMUSG000… Gnai3          243.      256.         240.       217.       84.7
2 ENSMUSG000… Pbsn             0         0            0          0         0
3 ENSMUSG000… Cdc45           11.2      13.8         11.6        4.27      8.35
4 ENSMUSG000… H19              6.31      8.53         7.09      11.0       0.194
5 ENSMUSG000… Scml2            2.19      4.66         2.80       2.50      1.24
6 ENSMUSG000… Apoh             0.224     0.0840       0          0         0
# ℹ 7 more variables: GSM1480296 &lt;dbl&gt;, GSM1480297 &lt;dbl&gt;, GSM1480298 &lt;dbl&gt;,
#   GSM1480299 &lt;dbl&gt;, GSM1480300 &lt;dbl&gt;, GSM1480301 &lt;dbl&gt;, GSM1480302 &lt;dbl&gt;</code></pre>
</div>
</div>
</div>
</div>
<p>We can also look at the last few lines with <code>tail()</code>. This
shows us the last 6 lines. This can be useful to check the bottom of the
file, that it looks ok.</p>
<div id="your-turn-2.6-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.6-1" class="callout-inner">
<h3 class="callout-title">Your turn 2.6</h3>
<div class="callout-content">
<p>Use <code>tail()</code> to look at the last few lines of
<code>counts</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tail</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 14
  ...1        gene_symbol GSM1480291 GSM1480292 GSM1480293 GSM1480294 GSM1480295
  &lt;chr&gt;       &lt;chr&gt;            &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 ENSMUSG000… Pcdha7           0.134     0          0            0          0
2 ENSMUSG000… Gm34240          0         0          0            0          0
3 ENSMUSG000… Pcdhga3          0.582     2.27       0.334        1.10       0
4 ENSMUSG000… Gm20750          0         0.0840     0.0417       0          0
5 ENSMUSG000… Rhbg             5.28      4.96       6.26         3.98       1.01
6 ENSMUSG000… Mat2a          212.      225.        97.9         70.1       22.0
# ℹ 7 more variables: GSM1480296 &lt;dbl&gt;, GSM1480297 &lt;dbl&gt;, GSM1480298 &lt;dbl&gt;,
#   GSM1480299 &lt;dbl&gt;, GSM1480300 &lt;dbl&gt;, GSM1480301 &lt;dbl&gt;, GSM1480302 &lt;dbl&gt;</code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-2.7-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-2.7-1" class="callout-inner">
<h3 class="callout-title">Your turn 2.7</h3>
<div class="callout-content">
<p>What are the cell types of the first 6 samples in the metadata for
the Fu et al. 2015 experiment?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 4
  ...1       characteristics               immunophenotype `developmental stage`
  &lt;chr&gt;      &lt;chr&gt;                         &lt;chr&gt;           &lt;chr&gt;
1 GSM1480291 mammary gland, luminal cells… luminal cell p… virgin
2 GSM1480292 mammary gland, luminal cells… luminal cell p… virgin
3 GSM1480293 mammary gland, luminal cells… luminal cell p… 18.5 day pregnancy
4 GSM1480294 mammary gland, luminal cells… luminal cell p… 18.5 day pregnancy
5 GSM1480295 mammary gland, luminal cells… luminal cell p… 2 day lactation
6 GSM1480296 mammary gland, luminal cells… luminal cell p… 2 day lactation      </code></pre>
</div>
<p>The first 6 samples are luminal cells.</p>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="dimensions-of-the-data">Dimensions of the data<a class="anchor" aria-label="anchor" href="#dimensions-of-the-data"></a>
</h2>
<hr class="half-width">
<p>You can print the number of rows and columns using the function
<code>dim()</code>.</p>
<p>For example:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 12  4</code></pre>
</div>
<p><code>sampleinfo</code> has 12 rows, corresponding to our 12 samples,
and 4 columns, corresponding to different features about the
samples.</p>
<div id="tip-always-verify-your-data-size" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="tip-always-verify-your-data-size" class="callout-inner">
<h3 class="callout-title">Tip: Always Verify Your Data Size</h3>
<div class="callout-content">
<p>Double-check that your data has the expected number of rows and
columns. It’s easy to read the wrong file or encounter corrupted
downloads. Catching these issues early will save you a lot of trouble
later!</p>
</div>
</div>
</div>
<div id="your-turn-2.8-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-2.8-1" class="callout-inner">
<h3 class="callout-title">Your turn 2.8</h3>
<div class="callout-content">
<p>Check how many rows and columns are in <code>counts</code>. Do these
numbers match what we expected?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23735    14</code></pre>
</div>
<p>We know that there are 12 samples in our data (see the diagram
above), and we don’t know how many genes were measured.</p>
<p><strong>Rows:</strong> 23735. This means there are 23735 genes, which
could be correct (we didn’t know how many genes were measured).</p>
<p><strong>Columns:</strong> 14. We would expect to have 12 columns
corresponding to our 12 samples, but instead we have 14. Why is this?
Have a look back at when we visualised the <code>counts</code> data,
there are two extra columns in the data corresponding to gene IDs and
gene names.</p>
</div>
</div>
</div>
</div>
<p>In the Environment Tab in the top right panel in RStudio we can also
see the number of rows and columns in the objects we have in our
session.</p>
<hr></section><section><h2 class="section-heading" id="column-and-row-names-of-the-data">Column and row names of the data<a class="anchor" aria-label="anchor" href="#column-and-row-names-of-the-data"></a>
</h2>
<hr class="half-width">
<div id="your-turn-2.9" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.9" class="callout-inner">
<h3 class="callout-title">Your turn 2.9</h3>
<div class="callout-content">
<p>Check the column and row names used in in <code>sampleinfo</code></p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="subsetting">Subsetting<a class="anchor" aria-label="anchor" href="#subsetting"></a>
</h2>
<hr class="half-width">
<p>Subsetting is very useful tool in R which allows you to extract parts
of the data you want to analyse. There are multiple ways to subset data
and here we’ll only cover a few.</p>
<p>We can use the <code>$</code> operator to access individual columns
by name.</p>
<div id="your-turn-2.10" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.10" class="callout-inner">
<h3 class="callout-title">Your turn 2.10</h3>
<div class="callout-content">
<p>Extract the ‘immunophenotype’ column of the metadata.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">immunophenotype</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>We can also use square brackets <code>[ ]</code> to access the rows
and columns of a matrix or a table.</p>
<p>For example, we can extract the first row ‘1’ of the data, using the
number on the <em>left-hand-side</em> of the comma.</p>
<div id="your-turn-2.11" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.11" class="callout-inner">
<h3 class="callout-title">Your turn 2.11</h3>
<div class="callout-content">
<p>Extract the first row using square brackets.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Here we extract the second column ‘2’ of the data, as indicated on
the <em>right-hand-side</em> of the comma.</p>
<div id="your-turn-2.12" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.12" class="callout-inner">
<h3 class="callout-title">Your turn 2.12</h3>
<div class="callout-content">
<p>Extract the second column using square brackets.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>You can use a combination of number of row and column to extract one
element in the matrix.</p>
<div id="your-turn-2.13" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.13" class="callout-inner">
<h3 class="callout-title">Your turn 2.13</h3>
<div class="callout-content">
<p>Extract the element in the first row and second column.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-2.14" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.14" class="callout-inner">
<h3 class="callout-title">Your turn 2.14</h3>
<div class="callout-content">
<p>We can also subset using a range of numbers. For example, if we
wanted the first three rows of <code>sampleinfo</code></p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span></span></code></pre>
</div>
<p>Or if we wanted the 2nd, 4th, and 5th row:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span>,<span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="the-c-function" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="the-c-function" class="callout-inner">
<h3 class="callout-title">The c() function</h3>
<div class="callout-content">
<p>We use the <code>c()</code> function extremely often in R when we
have multiple items that we are <em>combining</em> (‘c’ stands for
concatenating). We will see it again in this tutorial.</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="renaming-column-names">Renaming column names<a class="anchor" aria-label="anchor" href="#renaming-column-names"></a>
</h2>
<hr class="half-width">
<p>In the previous section, when we loaded in the data from the csv
file, we noticed that the first column had a missing column name and by
default, read_csv function assigned a name of “...1” to it. Let’s change
this column to something more descriptive now. We can do this by
combining a few things we’ve just learnt.</p>
<div id="your-turn-2.15" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.15" class="callout-inner">
<h3 class="callout-title">Your turn 2.15</h3>
<div class="callout-content">
<p>First, we use the <code>colnames()</code> function to obtain the
column names of sampleinfo. Then we use square brackets to subset the
first value of the column names (<code>[1]</code>). Last, we use the
assignment operator (<code>&lt;-</code>) to set the new value of the
first column name to “sample_id”.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"sample_id"</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Let’s check if this has been changed correctly.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 12 × 4
   sample_id  characteristics              immunophenotype `developmental stage`
   &lt;chr&gt;      &lt;chr&gt;                        &lt;chr&gt;           &lt;chr&gt;
 1 GSM1480291 mammary gland, luminal cell… luminal cell p… virgin
 2 GSM1480292 mammary gland, luminal cell… luminal cell p… virgin
 3 GSM1480293 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 4 GSM1480294 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 5 GSM1480295 mammary gland, luminal cell… luminal cell p… 2 day lactation
 6 GSM1480296 mammary gland, luminal cell… luminal cell p… 2 day lactation
 7 GSM1480297 mammary gland, basal cells,… basal cell pop… virgin
 8 GSM1480298 mammary gland, basal cells,… basal cell pop… virgin
 9 GSM1480299 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
10 GSM1480300 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
11 GSM1480301 mammary gland, basal cells,… basal cell pop… 2 day lactation
12 GSM1480302 mammary gland, basal cells,… basal cell pop… 2 day lactation      </code></pre>
</div>
<p>The first column is now named “sample_id”.</p>
<p>We can also do the same to the counts data. This time, we rename the
first column name from “...1” to “gene_id”.</p>
<div id="your-turn-2.16" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-2.16" class="callout-inner">
<h3 class="callout-title">Your turn 2.16</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"gene_id"</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Note: there are multiple ways to rename columns. We’ve covered one
way here, but another way is using the <code>rename()</code> function.
When programming, you’ll often find many ways to do the same thing.
Often there is one obvious method depending on the context you’re
in.</p>
<hr></section><section><h2 class="section-heading" id="structure-and-summary">Structure and Summary<a class="anchor" aria-label="anchor" href="#structure-and-summary"></a>
</h2>
<hr class="half-width">
<p>Other useful commands for checking data are <code>str()</code> and
<code>summary()</code>.</p>
<p><code>str()</code> shows us the structure of our data. It shows us
what columns there are, the first few entries, and what data type they
are e.g. character or numbers (double or integer).</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>spc_tbl_ [12 × 4] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ sample_id          : chr [1:12] "GSM1480291" "GSM1480292" "GSM1480293" "GSM1480294" ...
 $ characteristics    : chr [1:12] "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, 18.5 day pregnancy" "mammary gland, luminal cells, 18.5 day pregnancy" ...
 $ immunophenotype    : chr [1:12] "luminal cell population" "luminal cell population" "luminal cell population" "luminal cell population" ...
 $ developmental stage: chr [1:12] "virgin" "virgin" "18.5 day pregnancy" "18.5 day pregnancy" ...
 - attr(*, "spec")=
  .. cols(
  ..   ...1 = col_character(),
  ..   characteristics = col_character(),
  ..   immunophenotype = col_character(),
  ..   `developmental stage` = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
<p><code>summary()</code> generates summary statistics of our data. For
numeric columns (columns of type double or integer) it outputs
statistics such as the min, max, mean and median. We will demonstrate
this with the counts file as it contains numeric data. For character
columns it shows us the length (how many rows).</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   gene_id          gene_symbol          GSM1480291          GSM1480292
 Length:23735       Length:23735       Min.   :    0.000   Min.   :    0.000
 Class :character   Class :character   1st Qu.:    0.000   1st Qu.:    0.000
 Mode  :character   Mode  :character   Median :    1.745   Median :    1.891
                                       Mean   :   42.132   Mean   :   42.132
                                       3rd Qu.:   29.840   3rd Qu.:   29.604
                                       Max.   :12525.066   Max.   :12416.211
   GSM1480293          GSM1480294          GSM1480295
 Min.   :    0.000   Min.   :    0.000   Min.   :0.000e+00
 1st Qu.:    0.000   1st Qu.:    0.000   1st Qu.:0.000e+00
 Median :    0.918   Median :    0.888   Median :5.830e-01
 Mean   :   42.132   Mean   :   42.132   Mean   :4.213e+01
 3rd Qu.:   21.908   3rd Qu.:   19.921   3rd Qu.:1.227e+01
 Max.   :49191.148   Max.   :55692.086   Max.   :1.119e+05
   GSM1480296          GSM1480297          GSM1480298
 Min.   :0.000e+00   Min.   :    0.000   Min.   :    0.000
 1st Qu.:0.000e+00   1st Qu.:    0.000   1st Qu.:    0.000
 Median :5.440e-01   Median :    2.158   Median :    2.254
 Mean   :4.213e+01   Mean   :   42.132   Mean   :   42.132
 3rd Qu.:1.228e+01   3rd Qu.:   27.414   3rd Qu.:   26.450
 Max.   :1.087e+05   Max.   :10489.311   Max.   :10662.486
   GSM1480299          GSM1480300          GSM1480301
 Min.   :    0.000   Min.   :    0.000   Min.   :    0.000
 1st Qu.:    0.000   1st Qu.:    0.000   1st Qu.:    0.000
 Median :    1.854   Median :    1.816   Median :    1.629
 Mean   :   42.132   Mean   :   42.132   Mean   :   42.132
 3rd Qu.:   24.860   3rd Qu.:   23.443   3rd Qu.:   23.444
 Max.   :15194.048   Max.   :17434.935   Max.   :19152.728
   GSM1480302
 Min.   :    0.000
 1st Qu.:    0.000
 Median :    1.749
 Mean   :   42.132
 3rd Qu.:   24.818
 Max.   :15997.193  </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The <code>read_csv()</code> and <code>read_tsv()</code> functions
can be used to load in CSV and TSV files in R</li>
<li>The <code>head()</code> and <code>tail()</code> functions can print
the first and last parts of an object and the <code>dim()</code>
function prints the dimensions of an object</li>
<li>Subsetting can be done with the <code>$</code> operator using column
names or using square brackets <code>[ ]</code>
</li>
<li>The <code>str()</code> and <code>summary()</code> functions are
useful functions to get an overview or summary of the data</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-format-data"><p>Content from <a href="format-data.html">Formatting the Data</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/format-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What do the <code>pivot_longer()</code> and <code>full_join()</code>
functions do?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explore how to convert data frames from wide to long format</li>
<li>Join two data frames using column names</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="converting-from-wide-to-long-format">Converting from wide to long format<a class="anchor" aria-label="anchor" href="#converting-from-wide-to-long-format"></a>
</h2>
<hr class="half-width">
<p>We will first convert the data from wide format into long format to
make it easier to work with and so that we can plot the data with the
<code>ggplot2</code> package.</p>
<p>Instead of multiple columns with counts for each sample, we just want
one column containing all the expression values, as shown below:</p>
<figure><img src="fig/reshape_data.png" class="figure mx-auto d-block"></figure><p>We can use <code>pivot_longer()</code> to easily change the format
into long format.</p>
<div id="your-turn-3.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-3.1" class="callout-inner">
<h3 class="callout-title">Your turn 3.1</h3>
<div class="callout-content">
<p>Find out what <code>pivot_longer()</code> does and how to use it</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">pivot_longer</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-3.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-3.2" class="callout-inner">
<h3 class="callout-title">Your turn 3.2</h3>
<div class="callout-content">
<p>Convert the counts data into long format and save it as a new object
called <code>seqdata</code></p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">counts</span>, cols <span class="op">=</span> <span class="fu">starts_with</span><span class="op">(</span><span class="st">"GSM"</span><span class="op">)</span>, </span>
<span>                        names_to <span class="op">=</span> <span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code></pre>
</div>
<p>We use <code>cols = starts_with("GSM")</code> to tell the function we
want to reformat the columns whose names start with “GSM” (these columns
are the ones where we have the gene counts). <code>pivot_longer()</code>
will then reformat the specified columns into two new columns, which
we’re naming “Sample” and “Count”. The <code>names_to = "Sample"</code>
specifies that we want the new column containing the columns to be named
“Sample”, and the <code>values_to = "Count"</code> specifies that we
want the new column containing the values to be named “Count”.</p>
</div>
</div>
</div>
<p>As explained earlier, in R there is often more than one way to do the
same thing. We could get the same result by specifying the argument
<code>cols</code> in a different way. For example, instead of using
<code>starts_with</code> we could use a range like the one you used in
the previous section.</p>
<div id="your-turn-3.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-3.3" class="callout-inner">
<h3 class="callout-title">Your turn 3.3</h3>
<div class="callout-content">
<p>Convert the counts data into long format using a column range</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">counts</span>, cols <span class="op">=</span> <span class="va">GSM1480291</span><span class="op">:</span><span class="va">GSM1480302</span>, </span>
<span>                        names_to <span class="op">=</span> <span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Another way we could do the same thing is by specifying the columns
we <em>do not</em> want to reformat, this will tell
<code>pivot_longer()</code> to reformat all the other columns. To do
that we put <strong>a minus sign “-”</strong> in front of the column
names that we don’t want to reformat. This is a pretty common way to use
<code>pivot_longer()</code> as sometimes it is easier to exclude columns
we don’t want than include columns we do. The command below would give
us the same result as the previous command.</p>
<div id="your-turn-3.4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-3.4" class="callout-inner">
<h3 class="callout-title">Your turn 3.4</h3>
<div class="callout-content">
<p>Convert the counts data into long format by specifying which columns
not to convert</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">counts</span>, cols <span class="op">=</span> <span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="st">"gene_id"</span>, <span class="st">"gene_symbol"</span><span class="op">)</span>, </span>
<span>                        names_to <span class="op">=</span> <span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-3.5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-3.5" class="callout-inner">
<h3 class="callout-title">Your turn 3.5</h3>
<div class="callout-content">
<p>Type each command line above, then look at the data, are all three of
the <code>seqdata</code> objects you made the same?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<p>All three <code>seqdata</code> objects are the same, as we did the
same conversion from wide to long format using 3 different methods.
<code>seqdata</code> has 284820 rows and 4 columns, and has length
4.</p>
</div>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="joining-two-tables">Joining two tables<a class="anchor" aria-label="anchor" href="#joining-two-tables"></a>
</h2>
<hr class="half-width">
<p>Now that we’ve got just one column containing sample IDs in both our
counts and metadata objects we can join them together using the sample
IDs. This will make it easier to identify the categories for each sample
(e.g. if it’s basal cell type) and to use that information in our
plots.</p>
<figure><img src="fig/join_data.png" class="figure mx-auto d-block"></figure><p>We use the function <code>full_join()</code> and give as arguments
the two tables we want to join. We add
<code>by = join_by(Sample == sample_id)</code> to say we want to join on
the column called “Sample” in the first table (<code>seqdata</code>) and
the column called “sample_id” in the second table
(<code>sampleinfo</code>) when the values match:</p>
<div id="your-turn-3.6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-3.6" class="callout-inner">
<h3 class="callout-title">Your turn 3.6</h3>
<div class="callout-content">
<p>Join the count data and metadata by matching sample IDs</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allinfo</span> <span class="op">&lt;-</span> <span class="fu">full_join</span><span class="op">(</span><span class="va">seqdata</span>, <span class="va">sampleinfo</span>, by <span class="op">=</span> <span class="fu">join_by</span><span class="op">(</span><span class="va">Sample</span> <span class="op">==</span> <span class="va">sample_id</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-3.7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-3.7" class="callout-inner">
<h3 class="callout-title">Your turn 3.7</h3>
<div class="callout-content">
<p>Have a look at the new object you generated above and see what
information it includes, how many columns does it have and what does
each column tell you?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">allinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 284820      7</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">allinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "gene_id"             "gene_symbol"         "Sample"
[4] "Count"               "characteristics"     "immunophenotype"
[7] "developmental stage"</code></pre>
</div>
<p>The <code>allinfo</code> object has 7 columns which tell you about
the gene id and symbol, the same id, gene count, and sample information
(characteristics, immunophenotype, and developmental stage).</p>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>The <code>pivot_longer()</code> function can convert data frames
from wide to long format and there are multiple ways to do this</li>
<li>The <code>full_join()</code> function can merge two data frames. You
can specify which column names should be joined by using the
<code>join_by()</code> function.</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-ggplot2-intro"><p>Content from <a href="ggplot2-intro.html">Introduction to ggplot2</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/ggplot2-intro.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the three components needed for creating a plot in
ggplot2?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how to plot a basic plot in ggplot2</li>
<li>Learn how to modify plots using colours and facets</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="plotting-with-ggplot2">Plotting with <strong><code>ggplot2</code></strong>
<a class="anchor" aria-label="anchor" href="#plotting-with-ggplot2"></a>
</h2>
<hr class="half-width">
<p><strong><code>ggplot2</code></strong> is a plotting package that
makes it simple to create complex plots. One really great advantage
compared to classic R packages is that we only need to make minimal
changes if the underlying data change or if we decide to change our plot
type, for example, from a box plot to a violin plot. This helps in
creating publication quality plots with minimal amounts of adjustments
and tweaking.</p>
<p><strong><code>ggplot2</code></strong> likes data in the ‘long’
format, i.e., a column for every variable, and a row for every
observation, similar to what we created with <code>pivot_longer()</code>
above. Well-structured data will save you lots of time when making
figures with <strong><code>ggplot2</code></strong>.</p>
<div id="understanding-ggplot2-architecture" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<span class="callout-header">Callout</span>
<div id="understanding-ggplot2-architecture" class="callout-inner">
<h3 class="callout-title">Understanding ggplot2 Architecture</h3>
<div class="callout-content">
<p>ggplot2 plots are built step by step by adding layers with
<code>+</code>. This approach provides great flexibility, allowing you
to customize your plots extensively.</p>
</div>
</div>
</div>
<p>To build a ggplot, we use the following basic template that can be
used for different types of plots. Three things are required for a
ggplot:</p>
<figure><img src="fig/ggplot_template.png" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal">
<li>The data</li>
<li>The columns in the data we want to map to visual properties (called
aesthetics or <code>aes</code>) e.g. the columns for x values, y values
and colours</li>
<li>The type of plot (the <code>geom_</code>)</li>
</ol>
<p>There are different geoms we can use to create different types of
plot e.g. <code>geom_line()</code> <code>geom_point()</code>,
<code>geom_boxplot()</code>. To see the geoms available take a look at
the ggplot2 help or the handy <a href="https://rstudio.github.io/cheatsheets/data-visualization.pdf" class="external-link">ggplot2
cheatsheet</a>. Or if you type “geom” in RStudio, RStudio will show you
the different types of geoms you can use.</p>
<hr></section><section><h2 class="section-heading" id="creating-a-boxplot">Creating a boxplot<a class="anchor" aria-label="anchor" href="#creating-a-boxplot"></a>
</h2>
<hr class="half-width">
<p>Let’s plot boxplots to visualise the distribution of the counts for
each sample. This helps us to compare the samples and check if any look
unusual.</p>
<p><em>Note: In commands that span multiple lines in R, “+” must go at
the end of the line—it can’t go at the beginning.</em></p>
<div id="your-turn-4.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.1" class="callout-inner">
<h3 class="callout-title">Your turn 4.1</h3>
<div class="callout-content">
<p>Run the following command line. Identify the key functions
<code>aes</code> and type of plot:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-2-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<p>This plot looks a bit weird. It’s because we have some genes with
extremely high counts. To make it easier to visualise the distributions
we usually plot the logarithm of RNA-seq counts. We’ll plot the Sample
on the X axis and log2 Counts on the y axis. We can log the Counts
within the <code>aes()</code>. The sample labels are also overlapping
each other, we will show how to fix this later.</p>
<div id="your-turn-4.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.2" class="callout-inner">
<h3 class="callout-title">Your turn 4.2</h3>
<div class="callout-content">
<p>Generate a boxplot of log2 gene counts</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: Removed 84054 rows containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<p>We get a warning here about rows containing non-finite values being
removed. This is because some of the genes have a count of zero in the
samples and a log of zero is undefined. We can add +1 to every count to
avoid the zeros being dropped (‘psuedo-count’).</p>
<div id="your-turn-4.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.3" class="callout-inner">
<h3 class="callout-title">Your turn 4.3</h3>
<div class="callout-content">
<p>Generate a boxplot of log2 gene counts + 1</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<p>The box plots show that the distributions of the samples are not
identical but they are not very different.</p>
<hr></section><section><h2 class="section-heading" id="violin-plot">Violin plot<a class="anchor" aria-label="anchor" href="#violin-plot"></a>
</h2>
<hr class="half-width">
<p>Boxplots are useful summaries, but hide the shape of the
distribution. For example, if the distribution is bimodal, we would not
see it in a boxplot. An alternative to the boxplot is the <strong>violin
plot</strong>, where the shape (of the density of points) is drawn.</p>
<p>Let’s choose a different geom to do another type of plot.</p>
<div id="your-turn-4.4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-4.4" class="callout-inner">
<h3 class="callout-title">Your turn 4.4</h3>
<div class="callout-content">
<p>Using the same data (same x and y values), try editing the code above
to make a violin plot using the <code>geom_violin()</code> function.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plotting a violin plot</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_violin</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="colouring-by-categories">Colouring by categories<a class="anchor" aria-label="anchor" href="#colouring-by-categories"></a>
</h2>
<hr class="half-width">
<p>Let’s add a different colour for each sample.</p>
<div id="your-turn-4.5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.5" class="callout-inner">
<h3 class="callout-title">Your turn 4.5</h3>
<div class="callout-content">
<p>View the help file for <code>geom_boxplot</code> and scroll down to
<strong>Aesthetics</strong> heading. It specifies that there is an
option for colour.</p>
</div>
</div>
</div>
<div id="your-turn-4.6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.6" class="callout-inner">
<h3 class="callout-title">Your turn 4.6</h3>
<div class="callout-content">
<p>Map each sample to a colour using the <code>colour =</code> argument.
As we are mapping colour to a column in our data we need to put this
inside <code>aes()</code>.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<p>Colouring the edges wasn’t quite what we had in mind. Look at the
help for <code>geom_boxplot</code> to see what other aesthetic we could
use. Let’s try <code>fill =</code> instead.</p>
<div id="your-turn-4.7" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.7" class="callout-inner">
<h3 class="callout-title">Your turn 4.7</h3>
<div class="callout-content">
<p>Map each sample to a colour using the <code>fill =</code>
argument.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>That looks better. <code>fill =</code> is used to fill in areas in
ggplot2 plots, whereas <code>colour =</code> is used to colour lines and
points.</p>
</div>
</div>
</div>
<p>A really nice feature about ggplot is that we can easily colour by
another variable by simply changing the column we give to
<code>fill =</code>.</p>
<hr></section><section><h2 class="section-heading" id="creating-subplots-for-each-gene-using-faceting">Creating subplots for each gene using faceting<a class="anchor" aria-label="anchor" href="#creating-subplots-for-each-gene-using-faceting"></a>
</h2>
<hr class="half-width">
<p>With ggplot we can easily make subplots using <em>faceting</em>. For
example we can make <a href="http://www.sthda.com/english/wiki/ggplot2-stripchart-jitter-quick-start-guide-r-software-and-data-visualization" class="external-link">stripcharts</a>.
These are a type of scatterplot and are useful when there are a small
number of samples (when there are not too many points to visualise).
Here we will make stripcharts plotting expression by the groups (basal
virgin, basal pregnant, basal lactating, luminal virgin, luminal
pregnant, luminal lactating) for each gene.</p>
<div class="section level3">
<h3 id="shorter-category-names">Shorter category names<a class="anchor" aria-label="anchor" href="#shorter-category-names"></a>
</h3>
<p>As we saw in question 5.5, our column names are quite long, and this
might make them difficult to visualise on a plot. We can use the
function <code>mutate()</code> to add another column to our
<code>allinfo</code> object with shorter group names.</p>
<div id="your-turn-4.8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-4.8" class="callout-inner">
<h3 class="callout-title">Your turn 4.8</h3>
<div class="callout-content">
<p>Make a new column in <code>allinfo</code> with shortened category
names using the below code. How has the object <code>allinfo</code>
changed?</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allinfo</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">allinfo</span>, Group <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"basal.*virgin"</span><span class="op">)</span> <span class="op">~</span>  <span class="st">"bvirg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"basal.*preg"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"bpreg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"basal.*lact"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"blact"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"luminal.*virgin"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"lvirg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"luminal.*preg"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"lpreg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"luminal.*lact"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"llact"</span></span>
<span>       <span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Note: While not covered in this workshop, the above code uses <a href="https://en.wikipedia.org/wiki/Regular_expression" class="external-link">regular
expressions</a> to match patterns of characters in a string.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">allinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 8
  gene_id            gene_symbol Sample    Count characteristics immunophenotype
  &lt;chr&gt;              &lt;chr&gt;       &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;
1 ENSMUSG00000000001 Gnai3       GSM14802… 243.  mammary gland,… luminal cell p…
2 ENSMUSG00000000001 Gnai3       GSM14802… 256.  mammary gland,… luminal cell p…
3 ENSMUSG00000000001 Gnai3       GSM14802… 240.  mammary gland,… luminal cell p…
4 ENSMUSG00000000001 Gnai3       GSM14802… 217.  mammary gland,… luminal cell p…
5 ENSMUSG00000000001 Gnai3       GSM14802…  84.7 mammary gland,… luminal cell p…
6 ENSMUSG00000000001 Gnai3       GSM14802…  84.6 mammary gland,… luminal cell p…
# ℹ 2 more variables: `developmental stage` &lt;chr&gt;, Group &lt;chr&gt;</code></pre>
</div>
<p>We observe a new column called <code>Group</code> at the end which
has shortened category names, bvirg, lpreg, etc.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="filter-for-genes-of-interest">Filter for genes of interest<a class="anchor" aria-label="anchor" href="#filter-for-genes-of-interest"></a>
</h3>
<div id="your-turn-4.9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-4.9" class="callout-inner">
<h3 class="callout-title">Your turn 4.9</h3>
<div class="callout-content">
<p>How many genes are there in our data?</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3"> Show me the solution </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23735    14</code></pre>
</div>
<p>There are 23735 rows in our original counts data, so we have data on
23735 different genes. Note: we didn’t run <code>dim()</code> on our
<code>allinfo</code> object because this has multiple rows per gene.</p>
</div>
</div>
</div>
</div>
<p>Like our data set, most RNA-seq data sets have information on
thousands of genes, but most of them are usually not very interesting,
so we may want to filter them.</p>
<p>Here, we choose 8 genes with the highest counts summed across all
samples. They are listed here.</p>
<div id="your-turn-4.10" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.10" class="callout-inner">
<h3 class="callout-title">Your turn 4.10</h3>
<div class="callout-content">
<p>Create an object with a list of the 8 most highly expressed genes</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Csn1s2a"</span>, <span class="st">"Csn1s1"</span>, <span class="st">"Csn2"</span>, <span class="st">"Glycam1"</span>, <span class="st">"COX1"</span>, <span class="st">"Trf"</span>, <span class="st">"Wap"</span>, <span class="st">"Eef1a1"</span><span class="op">)</span></span></code></pre>
</div>
<p>We filter our data for just these genes of interest. We use
<code>%in%</code> to check if a value is in a set of values.</p>
<p>Filter the counts data to only include genes in the
<code>mygenes</code> object</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes_counts</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">allinfo</span>, <span class="va">gene_symbol</span> <span class="op">%in%</span> <span class="va">mygenes</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-4.11" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-4.11" class="callout-inner">
<h3 class="callout-title">Your turn 4.11</h3>
<div class="callout-content">
<p>Can you figure out how many rows <code>mygenes_counts</code> will
have without inspecting the object? Print the dimensions of the object
to check if you’re right.</p>
</div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4"> Show me the solution </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<p>There is one row per sample per gene in <code>mygenes_counts</code>
(as is the case in <code>allinfo</code>). As there are 8 genes left
after filtering, and 12 samples in our data, we expect there to be 96
rows in <code>mygenes_counts</code>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We expect there to be 8 x 12 rows in mygenes_counts</span></span>
<span><span class="fl">8</span> <span class="op">*</span> <span class="fl">12</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 96</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># That is correct!</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">mygenes_counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 96  8</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="accordionSpoiler1" class="accordion spoiler-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button spoiler-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSpoiler1" aria-expanded="false" aria-controls="collapseSpoiler1">
  <h3 class="accordion-header" id="headingSpoiler1">  <div class="note-square"><i aria-hidden="true" class="callout-icon" data-feather="eye"></i></div> Selecting the genes </h3>
</button>
<div id="collapseSpoiler1" class="accordion-collapse collapse" data-bs-parent="#accordionSpoiler1" aria-labelledby="headingSpoiler1">
<div class="accordion-body">
<p>To identify these 8 genes, we used <a href="https://r4ds.had.co.nz/pipes.html" class="external-link">pipes</a> (<code>%&gt;%</code>)
to string a series of function calls together (which is beyond the scope
of this tutorial, but totally worth learning about independently!).</p>
<pre><code><span><span class="va">mygenes</span> <span class="op">&lt;-</span> <span class="va">allinfo</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene_symbol</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>Total_count <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">Total_count</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">head</span><span class="op">(</span>n <span class="op">=</span> <span class="fl">8</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pull</span><span class="op">(</span><span class="va">gene_symbol</span><span class="op">)</span> </span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="faceting">Faceting<a class="anchor" aria-label="anchor" href="#faceting"></a>
</h3>
<div id="your-turn-4.12" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.12" class="callout-inner">
<h3 class="callout-title">Your turn 4.12</h3>
<div class="callout-content">
<p>Make boxplots faceted by gene, grouped and coloured by groups</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-14-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Here we <em>facet</em> on the <code>gene_symbol</code> column using
<code>facet_wrap()</code>. We add the tilde symbol <code>~</code> in
front of the column we want to facet on.</p>
</div>
</div>
</div>
<hr>
</div>
</section><section><h2 class="section-heading" id="scatterplots">Scatterplots<a class="anchor" aria-label="anchor" href="#scatterplots"></a>
</h2>
<hr class="half-width">
<p>In the example over, boxplots are not suitable because we only have
two values per group. Let’s plot the individual points instead using the
<code>geom_point()</code> to make a scatter plot.</p>
<div id="your-turn-4.13" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.13" class="callout-inner">
<h3 class="callout-title">Your turn 4.13</h3>
<div class="callout-content">
<p>Make scatter plots faceted by gene and grouped by groups</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-15-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<div class="section level3">
<h3 id="jitter-plot">Jitter plot<a class="anchor" aria-label="anchor" href="#jitter-plot"></a>
</h3>
<p>In the previous plots, the points are overlapping which makes it hard
to see them. We can make a jitter plot using <code>geom_jitter()</code>
which adds a small amount of random variation to the location of each
point so they do not overlap. If is also quite common to combine jitter
plots with other types of plot, for example, <a href="https://simplystatistics.org/2019/02/21/dynamite-plots-must-die/" class="external-link">jitter
with boxplot</a>.</p>
<div id="your-turn-4.14" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="your-turn-4.14" class="callout-inner">
<h3 class="callout-title">Your turn 4.14</h3>
<div class="callout-content">
<p>Make jitter plots faceted by gene and grouped by groups</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-16-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
<div id="your-turn-4.15" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<span class="callout-header">Challenge</span>
<div id="your-turn-4.15" class="callout-inner">
<h3 class="callout-title">Your turn 4.15</h3>
<div class="callout-content">
<p>Modify the code above to colour the jitter plots by group</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5"> Show me the solution </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-intro-rendered-unnamed-chunk-17-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Note that for jitter plots you will want to use the
<code>colour =</code> slot rather than the <code>fill =</code> slot.</p>
</div>
</div>
</div>
</div>
<hr>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>A ggplot has 3 components: data (dataset), mapping (columns to plot)
and geom (type of plot). Different types of plots include
<code>geom_point()</code>, <code>geom_jitter()</code>,
<code>geom_line()</code>, <code>geom_boxplot()</code>,
<code>geom_violin()</code>.</li>
<li>
<code>facet_wrap()</code> can be used to make subplots of the
data</li>
<li>The aesthetics of a ggplot can be modified, such as colouring by
different columns in the dataset</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div>
</section></section><section id="aio-ggplot2-customisation"><p>Content from <a href="ggplot2-customisation.html">Extra ggplot2 Customisation</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/ggplot2-customisation.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can you add a title and change the axis labels in a ggplot?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Understand how to specify colours of a plot, either manually or by
using a predefined palette</li>
<li>Explore modifying the <code>theme()</code> function in ggplot to
adjust plot aesthetics</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="specifying-colours">Specifying colours<a class="anchor" aria-label="anchor" href="#specifying-colours"></a>
</h2>
<hr class="half-width">
<p>You might want to control plotting colours. To see what colour names
are available you can type <code>colours()</code>. There is also an <a href="https://www.nceas.ucsb.edu/~frazier/RSpatialGuides/colorPaletteCheatsheet.pdf" class="external-link">R
colours cheatsheet</a> that shows what the colours look like.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mycolours</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"turquoise"</span>, <span class="st">"plum"</span>, <span class="st">"tomato"</span>, <span class="st">"violet"</span>, <span class="st">"steelblue"</span>, <span class="st">"chocolate"</span><span class="op">)</span></span></code></pre>
</div>
<p>Then we then add these colours to the plot using a <code>+</code> and
<code>scale_colour_manual(values = mycolours)</code>.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">mycolours</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-3-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are built-in colour palettes that can be handy to use, where
the sets of colours are predefined. <code>scale_colour_brewer()</code>
is a popular one (there is also <code>scale_fill_brewer()</code>). You
can take a look at the help for <code>scale_colour_brewer()</code> to
see what palettes are available. The <a href="https://www.nceas.ucsb.edu/~frazier/RSpatialGuides/colorPaletteCheatsheet.pdf" class="external-link">R
colours cheatsheet</a> also shows what the colours of the palettes look
like. There’s one called “Dark2”, let’s have a look at that.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_brewer</span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-4-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section><section><h2 class="section-heading" id="axis-labels-and-title">Axis labels and Title<a class="anchor" aria-label="anchor" href="#axis-labels-and-title"></a>
</h2>
<hr class="half-width">
<p>We can change the axis labels and add a title with
<code>labs()</code>. To change the x axis label we use
<code>labs(x = "New name")</code>. To change the y axis label we use
<code>labs(y = "New name")</code> or we can change them all at the same
time.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section><section><h2 class="section-heading" id="themes">Themes<a class="anchor" aria-label="anchor" href="#themes"></a>
</h2>
<hr class="half-width">
<p>We can adjust the text on the x axis (the group labels) by turning
them 90 degrees so we can read the labels better. To do this we modify
the ggplot theme. Themes are the non-data parts of the plot.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can remove the grey background and grid lines.</p>
<p>There are also a lot of built-in themes. Let’s have a look at a
couple of the more widely used themes. The default ggplot theme is
<code>theme_grey().</code></p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are many themes available, you can see some in the <a href="https://www.r-graph-gallery.com/192-ggplot-themes/" class="external-link">R graph
gallery</a>.</p>
<p>We can also modify parts of the theme individually. We can remove the
grey background and grid lines with the code below.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-9-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><hr></section><section><h2 class="section-heading" id="order-of-categories">Order of categories<a class="anchor" aria-label="anchor" href="#order-of-categories"></a>
</h2>
<hr class="half-width">
<p>The groups have been plotted in alphabetical order on the x axis and
in the legend (that is the default order), however, we may want to
change the order. We may prefer to plot the groups in order of stage,
for example, basal virgin, basal pregnant, basal lactate, luminal
virgin, luminal pregnant, luminal lactate.</p>
<p>First let’s make an object with the group order that we want.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">group_order</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"bvirg"</span>, <span class="st">"bpreg"</span>, <span class="st">"blact"</span>, <span class="st">"lvirg"</span>, <span class="st">"lpreg"</span>, <span class="st">"llact"</span><span class="op">)</span></span></code></pre>
</div>
<p>Next we need to make a column with the groups into an R data type
called a <strong>factor</strong>. Factors in R are a special data type
used to specify categories, you can read more about them in the <a href="https://r4ds.had.co.nz/vectors.html#augmented-vectors" class="external-link">R for Data
Science book</a>. The names of the categories are called the factor
<strong>levels</strong>.</p>
<p>We’ll add another column called “Group_f” where we’ll make the Group
column into a factor and specify what order we want the levels of the
factor.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes_counts</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">mygenes_counts</span>, Group_f <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Group</span>, levels <span class="op">=</span> <span class="va">group_order</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Take a look at the data. As the table is quite wide we can use
<code>select()</code> to select just the columns we want to view.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">mygenes_counts</span>, <span class="va">gene_id</span>, <span class="va">Group</span>, <span class="va">Group_f</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 96 × 3
   gene_id            Group Group_f
   &lt;chr&gt;              &lt;chr&gt; &lt;fct&gt;
 1 ENSMUSG00000000381 lvirg lvirg
 2 ENSMUSG00000000381 lvirg lvirg
 3 ENSMUSG00000000381 lpreg lpreg
 4 ENSMUSG00000000381 lpreg lpreg
 5 ENSMUSG00000000381 llact llact
 6 ENSMUSG00000000381 llact llact
 7 ENSMUSG00000000381 bvirg bvirg
 8 ENSMUSG00000000381 bvirg bvirg
 9 ENSMUSG00000000381 bpreg bpreg
10 ENSMUSG00000000381 bpreg bpreg
# ℹ 86 more rows</code></pre>
</div>
<p>Notice that the Group column has <code>&lt;chr&gt;</code> under the
heading, that indicates is a character data type, while the Group_f
column has <code>&lt;fct&gt;</code> under the heading, indicating it is
a factor data type. The <code>str()</code> command that we saw
previously is useful to check the data types in objects.</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">mygenes_counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>tibble [96 × 9] (S3: tbl_df/tbl/data.frame)
 $ gene_id            : chr [1:96] "ENSMUSG00000000381" "ENSMUSG00000000381" "ENSMUSG00000000381" "ENSMUSG00000000381" ...
 $ gene_symbol        : chr [1:96] "Wap" "Wap" "Wap" "Wap" ...
 $ Sample             : chr [1:96] "GSM1480291" "GSM1480292" "GSM1480293" "GSM1480294" ...
 $ Count              : num [1:96] 90.2 95.6 4140.3 8414.4 49204.9 ...
 $ characteristics    : chr [1:96] "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, 18.5 day pregnancy" "mammary gland, luminal cells, 18.5 day pregnancy" ...
 $ immunophenotype    : chr [1:96] "luminal cell population" "luminal cell population" "luminal cell population" "luminal cell population" ...
 $ developmental stage: chr [1:96] "virgin" "virgin" "18.5 day pregnancy" "18.5 day pregnancy" ...
 $ Group              : chr [1:96] "lvirg" "lvirg" "lpreg" "lpreg" ...
 $ Group_f            : Factor w/ 6 levels "bvirg","bpreg",..: 4 4 5 5 6 6 1 1 2 2 ...</code></pre>
</div>
<p><code>str()</code> shows us Group_f column is a Factor with 6 levels
(categories).</p>
<p>We can check the factor levels of a column as below.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">mygenes_counts</span><span class="op">$</span><span class="va">Group_f</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "bvirg" "bpreg" "blact" "lvirg" "lpreg" "llact"</code></pre>
</div>
<p>The levels are in the order that we want, so we can now change our
plot to use the “Group_f” column instead of Group column (change
<code>x =</code> and <code>colour =</code>).</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group_f</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group_f</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/ggplot2-customisation-rendered-unnamed-chunk-15-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We could do similar if we wanted to have the genes in the facets in a
different order. For example, we could add another column called
“gene_symbol_f” where we make the gene_symbol column into a factor,
specifying the order of the levels.</p>
<div id="exercise" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Make a colourblind-friendly plot using the colourblind-friendly
palettes <a href="https://www.cookbook-r.com/Graphs/Colors_(ggplot2)/#a-colorblind-friendly-palette" class="external-link">here</a>.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>In <code>ggplot2</code>, you can specify colours manually using the
<code>scale_colour_manual()</code> function or use a predefined palette
using the <code>scale_colour_brewer()</code> function</li>
<li>The <code>labs()</code> function allows you to set a plot title and
change axis labels</li>
<li>Complete themes such as <code>theme_bw()</code> and
<code>theme_classic()</code> can be used to change the appearance of a
plot</li>
<li>Using the <code>theme()</code> function allows you to tweak
components of a theme</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section><section id="aio-wrap-up"><p>Content from <a href="wrap-up.html">Wrapping Up</a></p>
<hr>
<p>Last updated on 2025-11-11 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/wrap-up.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can I save plots to a file?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Save plots to a pdf file in R</li>
<li>Explain why using the <code>sessionInfo()</code> function is good
practice</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="saving-plots">Saving plots<a class="anchor" aria-label="anchor" href="#saving-plots"></a>
</h2>
<hr class="half-width">
<p>We can save plots interactively by clicking Export in the Plots
window and saving as e.g. “myplot.pdf”. Or we can output plots to pdf
using <code>pdf()</code> followed by <code>dev.off()</code>. We put our
plot code after the call to <code>pdf()</code> and before closing the
plot device with <code>dev.off()</code>.</p>
<p>Let’s save our last plot.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pdf</span><span class="op">(</span><span class="st">"myplot.pdf"</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group_f</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group_f</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<hr></section><section><h2 class="section-heading" id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<p>At the end of your report, we recommend you run the
<code>sessionInfo()</code> function which prints out details about your
working environment such as the version of R yo are running, loaded
packages, and package versions. Printing out <code>sessionInfo()</code>
at the end of your analysis is good practice as it helps with
reproducibility in the future.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.5.2 (2025-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
 [1] lubridate_1.9.4 forcats_1.0.1   stringr_1.5.2   dplyr_1.1.4
 [5] purrr_1.1.0     readr_2.1.5     tidyr_1.3.1     tibble_3.3.0
 [9] ggplot2_4.0.0   tidyverse_2.0.0

loaded via a namespace (and not attached):
 [1] vctrs_0.6.5        cli_3.6.5          knitr_1.50         rlang_1.1.6
 [5] xfun_0.54          stringi_1.8.7      renv_1.1.5         generics_0.1.4
 [9] S7_0.2.0           glue_1.8.0         hms_1.1.4          scales_1.4.0
[13] grid_4.5.2         evaluate_1.0.5     tzdb_0.5.0         yaml_2.3.10
[17] lifecycle_1.0.4    compiler_4.5.2     RColorBrewer_1.1-3 timechange_0.3.0
[21] pkgconfig_2.0.3    farver_2.1.2       R6_2.6.1           tidyselect_1.2.1
[25] pillar_1.11.1      magrittr_2.0.4     tools_4.5.2        withr_3.0.2
[29] gtable_0.3.6      </code></pre>
</div>
<hr></section><section><h2 class="section-heading" id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<hr class="half-width">
<div id="exercise" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<span class="callout-header">Discussion</span>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Download the raw counts for this dataset from <a href="http://www.ilincs.org/apps/grein/" class="external-link">GREIN</a>.
<ol style="list-style-type: lower-alpha">
<li>Make a boxplot. Do the samples look any different to the normalised
counts?</li>
<li>Make subplots for the same set of 8 genes. Do they look any
different to the normalised counts?</li>
</ol>
</li>
<li>Download the normalised counts for the GSE63310 dataset from GREIN.
Make boxplots colouring the samples using different columns in the
metadata file.</li>
</ol>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="further-reading">Further Reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<hr class="half-width">
<ul>
<li>
<a href="https://pmacdasci.github.io/r-intro-tidyverse/" class="external-link">A short
intro to R and tidyverse</a><br>
</li>
<li>
<a href="https://r-statistics.co/Top50-Ggplot2-Visualizations-MasterList-R-Code.html" class="external-link">Top
50 Ggplot Visualisations</a><br>
</li>
<li><a href="https://r4ds.had.co.nz/" class="external-link">R for Data Science</a></li>
</ul>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<span class="callout-header">Key Points</span>
<div class="callout-inner">
<div class="callout-content">
<ul>
<li>You can use the <code>pdf()</code> function to save plots, and
finalize the file by calling <code>dev.off()</code>
</li>
<li>The <code>sessionInfo()</code> function prints information about
your R environment which is useful for reproducibility</li>
</ul>
</div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/melbournebioinformatics/intro-to-R/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/melbournebioinformatics/intro-to-R/" class="external-link">Source</a></p>
        <p>
        <a href="citation.html">Cite</a>
        | <a href="mailto:jchung@unimelb.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
		</div>
		<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors.</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.17.1" class="external-link">sandpaper (0.17.1)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>, and <a href="https://github.com/melbournebioinformatics/uom-varnish/tree/main" class="external-link">varnish (1.0.7)</a></p>
		</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "melbournebioinformatics.github.io/uom-varnish/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "R, programming, plotting, ggplot, tidyverse, data",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "melbournebioinformatics.github.io/uom-varnish/aio.html",
  "identifier": "melbournebioinformatics.github.io/uom-varnish/aio.html",
  "dateCreated": "2025-11-05",
  "dateModified": "2025-12-02",
  "datePublished": "2025-12-02"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

