<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>Lesson Title: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css">
<script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/uom/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicons/uom/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="../favicons/uom/favicon-16x16.png">
<link rel="manifest" href="../favicons/uom/site.webmanifest">
<link rel="mask-icon" href="../favicons/uom/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" media="(prefers-color-scheme: light)" content="white">
<meta name="theme-color" media="(prefers-color-scheme: dark)" content="black">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md top-nav uom"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="uom-logo" alt="Lesson Description" src="../assets/images/uom-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text">
<li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul>
</li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='../aio.html';">Learner View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav uom" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/uom-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      Lesson Title
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            Lesson Title
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul>
</li>
      </ul>
</div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Lesson Title
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress uom">
    <div class="progress-bar uom" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text">
<li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul>
</li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../aio.html">Learner View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->

            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction to R for Biologists</a>
    </div>
<!--/div.accordion-header-->

  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr>
<li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources">
<a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">

            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-introduction"><p>Content from <a href="introduction.html">Introduction to R for Biologists</a></p>
<hr>
<p>Last updated on 2025-02-04 |

        <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/edit/main/episodes/introduction.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 12 minutes</p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<p><strong>Authors:</strong> Maria Doyle, Jessica Chung, Vicky
Perreau.</p>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you write a lesson using R Markdown and
<a href="https://carpentries.github.io/sandpaper/" class="external-link">sandpaper</a>?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain how to use markdown with the new lesson template</li>
<li>Demonstrate how to include pieces of code, figures, and nested
challenge blocks</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<script>
function myFunction(id) {
    var x = document.getElementById(id);
    if (x.style.display === "none") {
        x.style.display = "block";
    } else {
        x.style.display = "none";
    }
}
</script><div class="section level1">
<h1 id="r-for-biologists-course">R for Biologists course<a class="anchor" aria-label="anchor" href="#r-for-biologists-course"></a>
</h1>
<p>R takes time to learn, like a spoken language. No one can expect to
be an R expert after learning R for a few hours. This course has been
designed to introduce biologists to R, showing some basics, and also
some powerful things R can do (things that would be more difficult to do
with Excel). The aim is to give beginners the confidence to continue
learning R, so the focus here is on tidyverse and visualisation of
biological data, as we believe this is a productive and engaging way to
start learning R. After this short introduction you could use <a href="https://bookdown.org/ansellbr/WEHI_tidyR_course_book" class="external-link">this
book</a> to dive a bit deeper.</p>
</div>
<div class="section level1">
<h1 id="intro-to-r-and-rstudio">Intro to R and RStudio<a class="anchor" aria-label="anchor" href="#intro-to-r-and-rstudio"></a>
</h1>
<p>RStudio is an interface that makes it easier to use R. There are four
windows in RStudio. The screenshot below shows an <a href="https://twitter.com/RLadiesNCL/status/1138812826917724160" class="external-link">analogy
linking the different RStudio windows to cooking</a>.</p>
<figure><img src="../fig/rstudio_cooking.jpg" class="figure mx-auto d-block"></figure><p>     </p>
<div class="section level2">
<h2 id="r-script-vs-console">R script vs console<a class="anchor" aria-label="anchor" href="#r-script-vs-console"></a>
</h2>
<p>There are two ways to work in RStudio in the console or in a script.
We can type a command in the console and press <code>Enter</code> to run
it. Try running the command below in the console.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">+</span> <span class="fl">1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2</code></pre>
</div>
<p>Or we can use an R script. To create a script, from the top menu in
RStudio: <code>File &gt; New File &gt; R Script</code>. Now type the
command below in the script. This time, to run the command, you use
<code>Ctrl</code> + <code>Enter</code> for Windows/Linux or
<code>Cmd</code> + <code>Enter</code> for MacOS. This sends the command
where the cursor is from the script to the console. You can highlight
multiple commands and then press <code>Cmd/Ctrl</code> +
<code>Enter</code> to run them one after the other.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">2</span> <span class="op">+</span> <span class="fl">2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
<p>As the RStudio screenshot above explains, if we work in the console
we don’t have a good record (recipe) of what we’ve done. We can see
commands we’ve run in the History panel (top right window), and we can
go backwards and forwards through our history in the console using the
up arrow and down arrow. But the history includes everything we’ve tried
to run, including our mistakes so it is good practice to use an R
script.</p>
<p>We can also add comments to a script. These are notes to ourself or
others about the commands in the script. Comments start with a
<code>#</code> which tells R not to run them as commands.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># testing R</span></span>
<span><span class="fl">2</span> <span class="op">+</span> <span class="fl">2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 4</code></pre>
</div>
<p>Keeping an accurate record of how you’ve manipulated your data is
important for <a href="https://ropensci.github.io/reproducibility-guide/sections/introduction/" class="external-link">reproducible
research</a>. Writing detailed comments and documenting your work are
useful reminders to your future self (and anyone else reading your
scripts) on what your code does.</p>
</div>
<div class="section level2">
<h2 id="working-directory">Working directory<a class="anchor" aria-label="anchor" href="#working-directory"></a>
</h2>
<p>Opening an RStudio session launches it from a specific location. This
is the ‘working directory’. <strong>R looks in the working directory by
default to read in data and save files.</strong> You can find out what
the working directory is by using the command <code>getwd()</code>. This
shows you the path to your working directory in the console. In Mac this
is in the format <code>/path/to/working/directory</code> and in Windows
<code>C:\path\to\working\directory</code>. It is often useful to have
your data and R scripts in the same directory and set this as your
working directory. We will do this now.</p>
<p>Make a folder for this course somewhere on your computer that you
will be able to easily find. Name the folder for example,
<code>Intro_R_course</code>. Then, to set this folder as your working
directory:</p>
<p>In RStudio click on the ‘Files’ tab and then click on the three dots,
as shown below.</p>
<figure><img src="../fig/three_dots.png" class="figure mx-auto d-block"></figure><p>In the window that appears, find the folder you created
(e.g. <code>Intro_R_course</code>), click on it, then click ‘Open’. The
files tab will now show the contents of your new folder. Click on
<code>More &gt; Set As Working Directory</code>, as shown below.</p>
<figure><img src="../fig/working_directory.png" class="figure mx-auto d-block"></figure><p>Save the script you created in the previous section as
<code>intro.R</code> in this directory. You can do this by clicking on
<code>File &gt; Save</code> and the default location should be the
current working directory (e.g. <code>Intro_R_course</code>).</p>
<p><em>Note: You can use an RStudio project as described <a href="https://r4ds.had.co.nz/workflow-projects.html" class="external-link">here</a> to
automatically keep track of and set the working directory.</em></p>
</div>
<div class="section level2">
<h2 id="packages">Packages<a class="anchor" aria-label="anchor" href="#packages"></a>
</h2>
<p>If it’s not already installed on your computer, you can use the
<code>install.packages</code> <strong>function</strong> to install a
<strong>package</strong>. A package is a collection of functions along
with documentation, code, tests and example data.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install.packages</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>We will see many functions in this tutorial. Functions are “canned
scripts” that automate more complicated sets of commands. Many functions
are predefined, or can be made available by importing R packages. A
function usually takes one or more inputs called <em>arguments</em>.
Here tidyverse is the argument to the <code>install.packages()</code>
function.</p>
<p><em>Note: functions require parentheses after the function
name.</em></p>
</div>
<div class="section level2">
<h2 id="getting-help">Getting help<a class="anchor" aria-label="anchor" href="#getting-help"></a>
</h2>
<p>To see what any function in R does, type a <code>?</code> before the
name and help information will appear in the Help panel on the right in
RStudio. Or you can search the function name in the Help panel search
box. Google and <a href="https://stackoverflow.com/" class="external-link">Stack Overflow</a>
are also useful resources for getting help.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">install.packages</span></span></code></pre>
</div>
<blockquote>
<h3 id="tab-completion">Tab completion</h3>
<p>A very useful feature is Tab completion. You can start typing and use
<kbd>Tab</kbd> to autocomplete code, for example, a function name.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="common-r-errors">Common R errors<a class="anchor" aria-label="anchor" href="#common-r-errors"></a>
</h2>
<p>R error messages are common and can sometimes be cryptic. You most
likely will encounter at least one error message during this tutorial.
Some common reasons for errors are:</p>
<ul>
<li>Case sensitivity. In R, as in other programming languages, case
sensitivity is important. ?install.packages is different to
?Install.packages.</li>
<li>Missing commas</li>
<li>Mismatched parentheses or brackets</li>
<li>Not quoting file paths</li>
<li>Not finishing a command so seeing “+” in the console. If you need
to, you can press ESC to cancel the command.</li>
</ul>
<p>To see examples of some R error messages with explanations see <a href="https://github.com/noamross/zero-dependency-problems/issues/7" class="external-link">here</a></p>
</div>
</div>
<div class="section level1">
<h1 id="getting-started-with-data">Getting started with data<a class="anchor" aria-label="anchor" href="#getting-started-with-data"></a>
</h1>
<div class="section level2">
<h2 id="data-files">Data files<a class="anchor" aria-label="anchor" href="#data-files"></a>
</h2>
<p>The data files required for this workshop are available on <a href="https://github.com/melbournebioinformatics/r-intro-biologists/blob/master/data.zip" class="external-link">GitHub</a>.
To download the data.zip file, you can click <a href="https://github.com/melbournebioinformatics/r-intro-biologists/raw/master/data.zip" class="external-link">here</a>.
Unzip the file and store this <code>data</code> folder in your working
directory.</p>
</div>
<div class="section level2">
<h2 id="grein-geo-rna-seq-experiments-interactive-navigator">GREIN (GEO RNA-seq Experiments Interactive Navigator)<a class="anchor" aria-label="anchor" href="#grein-geo-rna-seq-experiments-interactive-navigator"></a>
</h2>
<p>In this tutorial, we will learn some R through creating plots to
visualise data from an RNA-seq experiment. RNA-seq counts file can be
obtained from the <a href="https://www.nature.com/articles/s41598-019-43935-8" class="external-link">GREIN
platform</a>. GREIN provides &gt;6,500 published datasets from GEO that
have been uniformly processed. It is available at <a href="https://www.ilincs.org/apps/grein/" class="external-link uri">https://www.ilincs.org/apps/grein/</a>. You can search for a
dataset of interest using the GEO code. We obtained the dataset used
here using the code GSE60450. GREIN provide QC metrics for the RNA-seq
datasets and both raw and normalized counts. We will use the normalized
counts here. These are the counts of reads for each gene for each sample
normalized for differences in sequencing depth and composition bias.
Generally, the higher the number of counts the more the gene is
expressed.    </p>
</div>
<div class="section level2">
<h2 id="rna-seq-dataset">RNA-seq dataset<a class="anchor" aria-label="anchor" href="#rna-seq-dataset"></a>
</h2>
<p>Here we will create some plots using RNA-seq data from the paper by
<a href="https://www.ncbi.nlm.nih.gov/pubmed/25730472" class="external-link">Fu et
al. 2015</a>, GEO code GSE60450. This study examined expression in basal
and luminal cells from mice at different stages (virgin, pregnant and
lactating). There are 2 samples per group and 6 groups, 12 samples in
total.</p>
<figure><img src="../fig/mouse_exp.png" class="figure mx-auto d-block"></figure>
</div>
<div class="section level2">
<h2 id="tidyverse">Tidyverse<a class="anchor" aria-label="anchor" href="#tidyverse"></a>
</h2>
<figure><img src="../fig/tidyverse.png" style="width:100.0%" class="figure mx-auto d-block"></figure><p>In this course we will use the <strong>tidyverse</strong>. The
tidyverse is a collection of R packages that includes the extremely
widely used <strong><code>ggplot2</code></strong> package.</p>
<p>     </p>
<figure><img src="../fig/r_tidyverse.png" style="width:100.0%" class="figure mx-auto d-block"></figure><p>     </p>
<p>The <a href="https://www.tidyverse.org/" class="external-link">tidyverse</a> makes data
science faster, easier and more fun.</p>
<p>     </p>
<p><a href="https://www.openscapes.org/blog/2020/10/12/tidy-data/" class="external-link">Why
tidyverse? Why tidy data? Why is it such a game-changer?</a></p>
</div>
<div class="section level2">
<h2 id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h2>
<p>We use <code>library()</code> to load in the packages that we need.
As described in the cooking analogy in the first screenshot,
<code>install.packages()</code> is like buying a saucepan,
<code>library()</code> is taking it out of the cupboard to use it.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span></code></pre>
</div>
<p>The files we will use are csv comma-separated, so we will use the
<code>read_csv()</code> function from the tidyverse. There is also a
<code>read_tsv()</code> function for tab-separated values.</p>
<p>We will use the counts file called
<code>GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv</code> that’s
in a folder called <code>data</code> i.e. the path to the file should be
<code>data/GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv</code>.</p>
<p>We can read the counts file into R with the command below. We’ll
store the contents of the counts file in an <strong>object</strong>
called <code>counts</code>. This stores the file contents in R’s memory
making it easier to use.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in counts file</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>New names:
Rows: 23735 Columns: 14
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(2): ...1, gene_symbol dbl (12): GSM1480291, GSM1480292, GSM1480293,
GSM1480294, GSM1480295, GSM148...
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `` -&gt; `...1`</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in metadata</span></span>
<span><span class="va">sampleinfo</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/GSE60450_filtered_metadata.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>New names:
Rows: 12 Columns: 4
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(4): ...1, characteristics, immunophenotype, developmental stage
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `` -&gt; `...1`</code></pre>
</div>
<p>There is some information output by read_csv on “column
specification”. It tells us that there is a missing header and it has
been filled with the name “…1”. It also tells us what data types
read_csv is detecting in each column. Columns with text charactershave
been detected (col_character) and also columns with numbers
(col_double). We won’t get into the details of R data types in this
tutorial but they are important to know and you can read more about them
in the <a href="https://r4ds.had.co.nz/vectors.html#important-types-of-atomic-vector" class="external-link">R
for Data Science book</a>.</p>
<p>In R we use <code>&lt;-</code> to assign values to objects.
<code>&lt;-</code> is the <strong>assignment operator</strong>. It
assigns values on the right to objects on the left. So to create an
object, we need to give it a name (e.g. <code>counts</code>), followed
by the assignment operator <code>&lt;-</code>, and the value we want to
give it. We can give an object almost any name we want but there are
some rules and conventions as described in the <a href="https://style.tidyverse.org/syntax.html#object-names" class="external-link">tidyverse R
style guide</a></p>
<p>We can read in a file from a path on our computer on on the web and
use this as the value. Note that we need to put quotes (““) around file
paths.</p>
<div id="assignment-operator-shortcut" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="assignment-operator-shortcut" class="callout-inner">
<h3 class="callout-title">Assignment operator shortcut</h3>
<div class="callout-content">
<p>In RStudio, typing <kbd>Alt</kbd> + <kbd>-</kbd> (holding down
<kbd>Alt</kbd> at the same time as the <kbd>-</kbd> key) will write
<code>&lt;-</code> in a single keystroke in Windows, while typing &gt;
<kbd>Option</kbd> + <kbd>-</kbd> (holding down <kbd>Option</kbd> at the
same time as the <kbd>-</kbd> key) does the same in a Mac.</p>
</div>
</div>
</div>
<div id="exercises" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercises" class="callout-inner">
<h3 class="callout-title">Exercises</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li><p>Test what happens if you type
<code>Library(tidyverse)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type
<code>libary(tidyverse)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type
<code>library(tidyverse</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_tsv("data/GSE60450_filtered_metadata.csv")</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_csv("data/GSE60450_filtered_metadata.csv)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_csv("GSE60450_filtered_metadata.csv)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>What is the name of the first column you get with each of these 2
commands?<br><code>read.csv("data/GSE60450_filtered_metadata.csv")</code><br>
and<br><code>read_csv("data/GSE60450_filtered_metadata.csv")</code></p></li>
<li><p>If you run<br><code>read_csv("data/GSE60450_filtered_metadata.csv")</code><br>
what is the difference between the column header you see
<code>developmental stage</code> and ‘developmental stage’?</p></li>
</ol>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="getting-to-know-the-data">Getting to know the data<a class="anchor" aria-label="anchor" href="#getting-to-know-the-data"></a>
</h1>
<p>When assigning a value to an object, R does not print the value. For
example, here we don’t see what’s in the counts or sampleinfo files. But
there are ways we can look at the data. We will demonstrate using the
<code>sampleinfo</code> object.</p>
<p>We can type the name of the object and this will print the first few
lines and some information, such as number of rows.</p>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 12 × 4
   ...1       characteristics              immunophenotype `developmental stage`
   &lt;chr&gt;      &lt;chr&gt;                        &lt;chr&gt;           &lt;chr&gt;
 1 GSM1480291 mammary gland, luminal cell… luminal cell p… virgin
 2 GSM1480292 mammary gland, luminal cell… luminal cell p… virgin
 3 GSM1480293 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 4 GSM1480294 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 5 GSM1480295 mammary gland, luminal cell… luminal cell p… 2 day lactation
 6 GSM1480296 mammary gland, luminal cell… luminal cell p… 2 day lactation
 7 GSM1480297 mammary gland, basal cells,… basal cell pop… virgin
 8 GSM1480298 mammary gland, basal cells,… basal cell pop… virgin
 9 GSM1480299 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
10 GSM1480300 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
11 GSM1480301 mammary gland, basal cells,… basal cell pop… 2 day lactation
12 GSM1480302 mammary gland, basal cells,… basal cell pop… 2 day lactation      </code></pre>
</div>
<p>We can also use <code>dim()</code> to see the dimensions of an
object, the number of rows and columns.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 12  4</code></pre>
</div>
<p>This show us there are 12 rows and 4 columns.</p>
<p>In the Environment Tab in the top right panel in RStudio we can also
see the number of rows and columns in the objects we have in our
session.</p>
<p>We can also take a look the first few lines with <code>head()</code>.
This shows us the first 6 lines.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 4
  ...1       characteristics               immunophenotype `developmental stage`
  &lt;chr&gt;      &lt;chr&gt;                         &lt;chr&gt;           &lt;chr&gt;
1 GSM1480291 mammary gland, luminal cells… luminal cell p… virgin
2 GSM1480292 mammary gland, luminal cells… luminal cell p… virgin
3 GSM1480293 mammary gland, luminal cells… luminal cell p… 18.5 day pregnancy
4 GSM1480294 mammary gland, luminal cells… luminal cell p… 18.5 day pregnancy
5 GSM1480295 mammary gland, luminal cells… luminal cell p… 2 day lactation
6 GSM1480296 mammary gland, luminal cells… luminal cell p… 2 day lactation      </code></pre>
</div>
<p>We can look at the last few lines with <code>tail()</code>. This
shows us the last 6 lines. This can be useful to check the bottom of the
file, that it looks ok.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tail</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 4
  ...1       characteristics               immunophenotype `developmental stage`
  &lt;chr&gt;      &lt;chr&gt;                         &lt;chr&gt;           &lt;chr&gt;
1 GSM1480297 mammary gland, basal cells, … basal cell pop… virgin
2 GSM1480298 mammary gland, basal cells, … basal cell pop… virgin
3 GSM1480299 mammary gland, basal cells, … basal cell pop… 18.5 day pregnancy
4 GSM1480300 mammary gland, basal cells, … basal cell pop… 18.5 day pregnancy
5 GSM1480301 mammary gland, basal cells, … basal cell pop… 2 day lactation
6 GSM1480302 mammary gland, basal cells, … basal cell pop… 2 day lactation      </code></pre>
</div>
<p>Or we can see the whole file with <code>View()</code>.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<p>In the Environment tab we can see how many rows and columns the
object contains and we can click on the icon to view all the contents in
a tab. This runs the command View() for us.</p>
<p>We can see all the column names with <code>colnames()</code>.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "...1"                "characteristics"     "immunophenotype"
[4] "developmental stage"</code></pre>
</div>
<p>We can access individual columns by name using the <code>$</code>
symbol. For example we can see what’s contained in column …1.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">...1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "GSM1480291" "GSM1480292" "GSM1480293" "GSM1480294" "GSM1480295"
 [6] "GSM1480296" "GSM1480297" "GSM1480298" "GSM1480299" "GSM1480300"
[11] "GSM1480301" "GSM1480302"</code></pre>
</div>
<p>If we just wanted to see the first 3 values in the column we can
specify this using square brackets.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">...1</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "GSM1480291" "GSM1480292" "GSM1480293"</code></pre>
</div>
<p>Other useful commands for checking data are <code>str()</code> and
<code>summary()</code>.</p>
<p><code>str()</code> shows us the structure of our data. It shows us
what columns there are, the first few entries, and what data type they
are e.g. character or numbers (double or integer).</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>spc_tbl_ [12 × 4] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ ...1               : chr [1:12] "GSM1480291" "GSM1480292" "GSM1480293" "GSM1480294" ...
 $ characteristics    : chr [1:12] "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, 18.5 day pregnancy" "mammary gland, luminal cells, 18.5 day pregnancy" ...
 $ immunophenotype    : chr [1:12] "luminal cell population" "luminal cell population" "luminal cell population" "luminal cell population" ...
 $ developmental stage: chr [1:12] "virgin" "virgin" "18.5 day pregnancy" "18.5 day pregnancy" ...
 - attr(*, "spec")=
  .. cols(
  ..   ...1 = col_character(),
  ..   characteristics = col_character(),
  ..   immunophenotype = col_character(),
  ..   `developmental stage` = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
<p><code>summary()</code> generates summary statistics of our data. For
numeric columns (columns of type double or integer) it outputs
statistics such as the min, max, mean and median. We will demonstrate
this with the counts file as it contains numeric data. For character
columns it shows us the length (how many rows).</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     ...1           gene_symbol          GSM1480291          GSM1480292
 Length:23735       Length:23735       Min.   :    0.000   Min.   :    0.000
 Class :character   Class :character   1st Qu.:    0.000   1st Qu.:    0.000
 Mode  :character   Mode  :character   Median :    1.745   Median :    1.891
                                       Mean   :   42.132   Mean   :   42.132
                                       3rd Qu.:   29.840   3rd Qu.:   29.604
                                       Max.   :12525.066   Max.   :12416.211
   GSM1480293         GSM1480294         GSM1480295          GSM1480296
 Min.   :    0.00   Min.   :    0.00   Min.   :     0.00   Min.   :     0.00
 1st Qu.:    0.00   1st Qu.:    0.00   1st Qu.:     0.00   1st Qu.:     0.00
 Median :    0.92   Median :    0.89   Median :     0.58   Median :     0.54
 Mean   :   42.13   Mean   :   42.13   Mean   :    42.13   Mean   :    42.13
 3rd Qu.:   21.91   3rd Qu.:   19.92   3rd Qu.:    12.27   3rd Qu.:    12.28
 Max.   :49191.15   Max.   :55692.09   Max.   :111850.87   Max.   :108726.08
   GSM1480297          GSM1480298          GSM1480299
 Min.   :    0.000   Min.   :    0.000   Min.   :    0.000
 1st Qu.:    0.000   1st Qu.:    0.000   1st Qu.:    0.000
 Median :    2.158   Median :    2.254   Median :    1.854
 Mean   :   42.132   Mean   :   42.132   Mean   :   42.132
 3rd Qu.:   27.414   3rd Qu.:   26.450   3rd Qu.:   24.860
 Max.   :10489.311   Max.   :10662.486   Max.   :15194.048
   GSM1480300          GSM1480301          GSM1480302
 Min.   :    0.000   Min.   :    0.000   Min.   :    0.000
 1st Qu.:    0.000   1st Qu.:    0.000   1st Qu.:    0.000
 Median :    1.816   Median :    1.629   Median :    1.749
 Mean   :   42.132   Mean   :   42.132   Mean   :   42.132
 3rd Qu.:   23.443   3rd Qu.:   23.443   3rd Qu.:   24.818
 Max.   :17434.935   Max.   :19152.728   Max.   :15997.193  </code></pre>
</div>
</div>
<div class="section level1">
<h1 id="formatting-the-data">Formatting the data<a class="anchor" aria-label="anchor" href="#formatting-the-data"></a>
</h1>
<div class="section level2">
<h2 id="converting-from-wide-to-long-format">Converting from wide to long format<a class="anchor" aria-label="anchor" href="#converting-from-wide-to-long-format"></a>
</h2>
<p>We will first convert the data from wide format into long format to
make it easier to work with and plot with ggplot. We want just one
column containing all the expression values instead of multiple columns
with counts for each sample, as shown in the image below.</p>
<figure><img src="../fig/reshape_data.png" class="figure mx-auto d-block"></figure><p>We can use <code>pivot_longer()</code> to easily change the format
into long format.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">counts</span>, cols <span class="op">=</span> <span class="fu">starts_with</span><span class="op">(</span><span class="st">"GSM"</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"Sample"</span>, </span>
<span>                        values_to <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code></pre>
</div>
<p>We use <code>cols = starts_with("GSM")</code> to tell the function we
want to reformat the columns whose names start with “GSM”.
<code>pivot_longer()</code> will then reformat the specified columns
into two new columns, which we’re naming “Sample” and “Count”. The
<code>names_to = "Sample"</code> specifies that we want the new column
containing the columns we sepcified with <code>cols</code> to be named
“Sample”, and the <code>values_to = "Count"</code> specifies that we
want the new column contining the values to be named “Count”.</p>
<p>We could also specify a column range to reformat. The command below
would give us the same result as the previous command.</p>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">counts</span>, cols <span class="op">=</span> <span class="va">GSM1480291</span><span class="op">:</span><span class="va">GSM1480302</span>, </span>
<span>                        names_to <span class="op">=</span> <span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code></pre>
</div>
<p>Alternatively, we could specify the columns we <em>don’t</em> want to
reformat and <code>pivot_longer()</code> will reformat all the other
columns. To do that we put a minus sign “-” in front of the column names
that we don’t want to reformat. This is a pretty common way to use
<code>pivot_longer()</code> as sometimes it is easier to exclude columns
we don’t want than include columns we do. The command below would give
us the same result as the previous command.</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span> <span class="op">&lt;-</span> <span class="fu">pivot_longer</span><span class="op">(</span><span class="va">counts</span>, cols <span class="op">=</span> <span class="op">-</span><span class="fu">c</span><span class="op">(</span><span class="st">"...1"</span>, <span class="st">"gene_symbol"</span><span class="op">)</span>, </span>
<span>                        names_to <span class="op">=</span> <span class="st">"Sample"</span>, values_to <span class="op">=</span> <span class="st">"Count"</span><span class="op">)</span></span></code></pre>
</div>
<p>Here we see the function <code>c()</code> for the first time. We use
this function extremely often in R when we have multiple items that we
are <em>combining</em>. We will see it again in this tutorial.</p>
<p>Let’s have a look at the data.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seqdata</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 284,820 × 4
   ...1               gene_symbol Sample     Count
   &lt;chr&gt;              &lt;chr&gt;       &lt;chr&gt;      &lt;dbl&gt;
 1 ENSMUSG00000000001 Gnai3       GSM1480291 243.
 2 ENSMUSG00000000001 Gnai3       GSM1480292 256.
 3 ENSMUSG00000000001 Gnai3       GSM1480293 240.
 4 ENSMUSG00000000001 Gnai3       GSM1480294 217.
 5 ENSMUSG00000000001 Gnai3       GSM1480295  84.7
 6 ENSMUSG00000000001 Gnai3       GSM1480296  84.6
 7 ENSMUSG00000000001 Gnai3       GSM1480297 175.
 8 ENSMUSG00000000001 Gnai3       GSM1480298 187.
 9 ENSMUSG00000000001 Gnai3       GSM1480299 177.
10 ENSMUSG00000000001 Gnai3       GSM1480300 169.
# ℹ 284,810 more rows</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="joining-two-tables">Joining two tables<a class="anchor" aria-label="anchor" href="#joining-two-tables"></a>
</h2>
<p>Now that we’ve got just one column containing sample ids in both our
counts and metadata objects we can join them together using the sample
ids. This will make it easier to identify the categories for each sample
(e.g. if it’s basal cell type) and to use that information in our
plots.</p>
<figure><img src="../fig/join_data.png" class="figure mx-auto d-block"></figure><p>We will use the function <code>full_join()</code> and give it the two
tables we want to join. We add <code>by = c("Sample" = "...1")</code> to
say we want to join on the column called “Sample”” in the first table
(<code>seqdata</code>) and the column called “…1” in the second table
(<code>sampleinfo</code>)</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allinfo</span> <span class="op">&lt;-</span> <span class="fu">full_join</span><span class="op">(</span><span class="va">seqdata</span>, <span class="va">sampleinfo</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Sample"</span> <span class="op">=</span> <span class="st">"...1"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s have a look at the data.</p>
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allinfo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 284,820 × 7
   ...1               gene_symbol Sample   Count characteristics immunophenotype
   &lt;chr&gt;              &lt;chr&gt;       &lt;chr&gt;    &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;
 1 ENSMUSG00000000001 Gnai3       GSM1480… 243.  mammary gland,… luminal cell p…
 2 ENSMUSG00000000001 Gnai3       GSM1480… 256.  mammary gland,… luminal cell p…
 3 ENSMUSG00000000001 Gnai3       GSM1480… 240.  mammary gland,… luminal cell p…
 4 ENSMUSG00000000001 Gnai3       GSM1480… 217.  mammary gland,… luminal cell p…
 5 ENSMUSG00000000001 Gnai3       GSM1480…  84.7 mammary gland,… luminal cell p…
 6 ENSMUSG00000000001 Gnai3       GSM1480…  84.6 mammary gland,… luminal cell p…
 7 ENSMUSG00000000001 Gnai3       GSM1480… 175.  mammary gland,… basal cell pop…
 8 ENSMUSG00000000001 Gnai3       GSM1480… 187.  mammary gland,… basal cell pop…
 9 ENSMUSG00000000001 Gnai3       GSM1480… 177.  mammary gland,… basal cell pop…
10 ENSMUSG00000000001 Gnai3       GSM1480… 169.  mammary gland,… basal cell pop…
# ℹ 284,810 more rows
# ℹ 1 more variable: `developmental stage` &lt;chr&gt;</code></pre>
</div>
<p>The two tables have been joined.</p>
<div id="exercise" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>View the help page of the <code>head</code> function and find the
“Arguments” heading. What does the <code>n</code> argument do? How many
rows and columns do you get with
<code>head(sampleinfo, n = 8)</code>?<br>
</li>
<li>Store the output of the first 20 lines of the <code>counts</code>
object in a new variable named <code>subset_counts</code>. What is the
gene_symbol in row 20?<br>
</li>
<li>View the values in the <code>GSM1480291</code> column from your
<code>subset_counts</code> object using the <code>$</code> subsetting
method. What is the 5th value?<br>
</li>
<li>View the help page of the <code>mean</code> function. What is the
mean of the column of values you obtained from the previous
question?<br>
</li>
<li>How can you use <code>pivot_longer</code> to transform
<code>dat</code> into a ‘tidy’ data called <code>dat_long</code> that
contains 3 columns (sample, experiment, count).</li>
</ol>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat</span> <span class="op">&lt;-</span>  <span class="fu">tibble</span><span class="op">(</span>sample <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span>             experiment_1 <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>             experiment_2 <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p><code>dat_long</code> should look similar to what you get if you
paste this into the console (the values in the count column will be
different):</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_long</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>experiment <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="st">"experiment_1"</span>, <span class="st">"experiment_2"</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span>count <span class="op">=</span> <span class="fu">rnorm</span><span class="op">(</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dat_long</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 20 × 3
   sample experiment     count
    &lt;int&gt; &lt;chr&gt;          &lt;dbl&gt;
 1      1 experiment_1  0.937
 2      1 experiment_2  1.63
 3      2 experiment_1 -0.345
 4      2 experiment_2 -0.0334
 5      3 experiment_1  1.27
 6      3 experiment_2 -0.841
 7      4 experiment_1 -1.68
 8      4 experiment_2  0.143
 9      5 experiment_1  0.0443
10      5 experiment_2  1.23
11      6 experiment_1 -0.118
12      6 experiment_2  0.417
13      7 experiment_1 -0.160
14      7 experiment_2  0.431
15      8 experiment_1 -1.18
16      8 experiment_2 -0.283
17      9 experiment_1  0.282
18      9 experiment_2  0.152
19     10 experiment_1  0.692
20     10 experiment_2 -1.28  </code></pre>
</div>
<ol start="6" style="list-style-type: decimal">
<li>If you have another table with sample information such as</li>
</ol>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span>sample <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,</span>
<span>group <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"Mutant"</span>, <span class="fl">50</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="st">"Control"</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Join <code>dat_long</code> to <code>sampleinfo</code> using the
common column called sample. How many rows do you get if you use i)
<code>full_join</code>, ii) <code>left_join</code>, iii)
<code>right_join</code>, iv) <code>inner_join</code>?</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="plotting-with-ggplot2">Plotting with <strong><code>ggplot2</code></strong>
<a class="anchor" aria-label="anchor" href="#plotting-with-ggplot2"></a>
</h1>
<p><strong><code>ggplot2</code></strong> is a plotting package that
makes it simple to create complex plots. One really great benefit of
ggplot2 versus the older base R plotting is that we only need to make
minimal changes if the underlying data change or if we decide to change
our plot type, for example, from a box plot to a violin plot. This helps
in creating publication quality plots with minimal amounts of
adjustments and tweaking.</p>
<p><strong><code>ggplot2</code></strong> likes data in the ‘long’
format, i.e., a column for every variable, and a row for every
observation, similar to what we created with
<code>pivot_longer()</code>. Well-structured data will save you lots of
time when making figures with <strong><code>ggplot2</code></strong>.</p>
<p>As we shall see, ggplot graphics are built step by step by adding new
elements using the <code>+</code>. Adding layers in this fashion allows
for extensive flexibility and customization of plots.</p>
<p>To build a ggplot, we use the following basic template that can be
used for different types of plots. Three things are required for a
ggplot:</p>
<figure><img src="../fig/ggplot_template.png" class="figure mx-auto d-block"></figure><ol style="list-style-type: decimal">
<li>The data</li>
<li>The columns in the data we want to map to visual properties (called
aesthetics or aes in ggplot2) e.g. the columns for x values, y values
and colours</li>
<li>The type of plot (the geom_)</li>
</ol>
<p>There are different geoms we can use to create different types of
plot e.g. <code>geom_line()</code> <code>geom_point()</code>,
<code>geom_boxplot()</code>. To see the geoms available take a look at
the ggplot2 help or the handy <a href="https://github.com/rstudio/cheatsheets/raw/master/data-visualization-2.1.pdf" class="external-link">ggplot2
cheatsheet</a>. Or if you type “geom” in RStudio, RStudio will show you
the different types of geoms you can use.</p>
<div class="section level2">
<h2 id="creating-a-boxplot">Creating a boxplot<a class="anchor" aria-label="anchor" href="#creating-a-boxplot"></a>
</h2>
<p>We can make boxplots to visualise the distribution of the counts for
each sample. This helps us to compare the samples and check if any look
unusual.</p>
<p><em>Note: with ggplot the “+” must go at the end of the line, it
can’t go at the beginning.</em></p>
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-27-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We have generated our first plot!</p>
<p>But it looks a bit weird. It’s because we have some genes with
extremely high counts. To make it easier to visualise the distributions
we usually plot the logarithm of RNA-seq counts. We’ll plot the Sample
on the X axis and log~ 2~ Counts on the y axis. We can log the Counts
within the <code>aes()</code>. The sample labels are also overlapping
each other, we will show how to fix this later.</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: Removed 84054 rows containing non-finite outside the scale range
(`stat_boxplot()`).</code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-28-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We get a warning here about rows containing non-finite values being
removed. This is because some of the genes have a count of zero in the
samples and a log of zero is undefined. We can add a small number to
every count to avoid the zeros being dropped.</p>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-29-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The box plots show that the distributions of the samples are not
identical but they are not very different.</p>
<p>Box plots are useful summaries, but hide the shape of the
distribution. For example, if the distribution is bimodal, we would not
see it in a boxplot. An alternative to the boxplot is the <strong>violin
plot</strong>, where the shape (of the density of points) is drawn. See
<a href="https://blog.bioturing.com/2018/05/16/5-reasons-you-should-use-a-violin-graph/" class="external-link">here</a>
for an example of how differences in distribution may be hidden in box
plots but revealed with violin plots.</p>
<div id="exercise-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-1" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>You can easily make different types of plots with ggplot by using
different geoms. Using the same data (same x and y values), try editing
the code above to make a violin plot (Hint: there’s a
<code>geom_violin</code>)</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="colouring-by-categories">Colouring by categories<a class="anchor" aria-label="anchor" href="#colouring-by-categories"></a>
</h1>
<p>What if we would like to add some colour to the plot, for example, a
different colour for each sample.</p>
<p>If we look at the <code>geom_boxplot</code> help we can see under the
heading called “Aesthetics” that there’s an option for colour. Let’s try
adding that to our plot. We’ll specify we want to map the Sample column
to <code>colour =</code>. As we are mapping colour to a column in our
data we need to put this inside the <code>aes()</code>.</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-30-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Colouring the edges wasn’t quite what we had in mind. Look at the
help for <code>geom_boxplot</code> to see what other aesthetic we could
use. Let’s try <code>fill =</code> instead.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">allinfo</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Sample</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">Sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-31-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>That looks better. <code>fill =</code> is used to
<strong>fill</strong> in areas in ggplot2 plots, whereas
<code>colour =</code> is used to colour lines and points.</p>
<p>A really nice feature about ggplot is that we can easily colour by
another variable by simply changing the column we give to
<code>fill =</code>.</p>
<div id="exercise-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-2" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<p>Modify the plot above. Colour by other variables (columns) in the
metadata file:</p>
<ol style="list-style-type: decimal">
<li>characteristics</li>
<li>immunophenotype</li>
<li>`developmental stage` (As there is a space in the column name we
need to use backticks around the name (``).</li>
</ol>
<p><em>Note: backticks are not single quotes (''). The backtick key is
usually at the top left corner of a laptop keyboard under the ESC key.
Check what happens if you don’t use backticks.)</em></p>
<div class="section level4">
<h4 id="optional-exercise">Optional exercise<a class="anchor" aria-label="anchor" href="#optional-exercise"></a>
</h4>
<p>The <code>geom_boxplot</code> function can also take in additional
arguments. For example, you can decrease the size of the outlier points
by using the <code>outlier.size</code> argument like so:
<code>geom_boxplot(outlier.size = 0.5)</code>. View the help page for
<code>geom_boxplot</code>. Can you find a way to hide outliers
altogether? Plot a boxplot with hidden outliers.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="creating-subplots-for-each-gene">Creating subplots for each gene<a class="anchor" aria-label="anchor" href="#creating-subplots-for-each-gene"></a>
</h1>
<p>With ggplot we can easily make subplots using <em>faceting</em>. For
example we can make <a href="https://www.sthda.com/english/wiki/ggplot2-stripchart-jitter-quick-start-guide-r-software-and-data-visualization" class="external-link">stripcharts</a>.
These are a type of scatterplot and are useful when there are a small
number of samples (when there are not too many points to visualise).
Here we will make stripcharts plotting expression by the groups (basal
virgin, basal pregnant, basal lactating, luminal virgin, luminal
pregnant, luminal lactating) for each gene.</p>
<div class="section level2">
<h2 id="make-shorter-category-names">Make shorter category names<a class="anchor" aria-label="anchor" href="#make-shorter-category-names"></a>
</h2>
<p>First we’ll use <code>mutate()</code> to add a column with shorter
group names to use in the plot, as the group names in the
characteristics column are quite long.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allinfo</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">allinfo</span>, Group <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"basal.*virgin"</span><span class="op">)</span> <span class="op">~</span>  <span class="st">"bvirg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"basal.*preg"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"bpreg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"basal.*lact"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"blact"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"luminal.*virgin"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"lvirg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"luminal.*preg"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"lpreg"</span>,</span>
<span>        <span class="fu">str_detect</span><span class="op">(</span><span class="va">characteristics</span>, <span class="st">"luminal.*lact"</span><span class="op">)</span>  <span class="op">~</span>  <span class="st">"llact"</span></span>
<span>       <span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Have a look at this data using <code>head()</code>. You should see a
new column called <code>Group</code> has been added to the end.</p>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">allinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 8
  ...1               gene_symbol Sample    Count characteristics immunophenotype
  &lt;chr&gt;              &lt;chr&gt;       &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;
1 ENSMUSG00000000001 Gnai3       GSM14802… 243.  mammary gland,… luminal cell p…
2 ENSMUSG00000000001 Gnai3       GSM14802… 256.  mammary gland,… luminal cell p…
3 ENSMUSG00000000001 Gnai3       GSM14802… 240.  mammary gland,… luminal cell p…
4 ENSMUSG00000000001 Gnai3       GSM14802… 217.  mammary gland,… luminal cell p…
5 ENSMUSG00000000001 Gnai3       GSM14802…  84.7 mammary gland,… luminal cell p…
6 ENSMUSG00000000001 Gnai3       GSM14802…  84.6 mammary gland,… luminal cell p…
# ℹ 2 more variables: `developmental stage` &lt;chr&gt;, Group &lt;chr&gt;</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="filter-for-genes-of-interest">Filter for genes of interest<a class="anchor" aria-label="anchor" href="#filter-for-genes-of-interest"></a>
</h2>
<p>We can make plots for a set of given genes.</p>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Csn1s2a"</span>, <span class="st">"Csn1s1"</span>, <span class="st">"Csn2"</span>, <span class="st">"Glycam1"</span>, <span class="st">"CO...1"</span>, <span class="st">"Trf"</span>, <span class="st">"Wap"</span>, <span class="st">"Eef1a1"</span><span class="op">)</span></span></code></pre>
</div>
<p>We filter our data for just these genes of interest. We use
<code>%in%</code> to check if a value is in a set of values.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes_counts</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">allinfo</span>, <span class="va">gene_symbol</span> <span class="op">%in%</span> <span class="va">mygenes</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="create-plots-for-each-gene">Create plots for each gene<a class="anchor" aria-label="anchor" href="#create-plots-for-each-gene"></a>
</h2>
<p>We can make boxplots for just these genes. We <em>facet</em> on the
<code>gene_symbol</code> column using <code>facet_wrap()</code>. We add
the tilde symbol <code>~</code> in front of the column we want to facet
on.</p>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-36-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The boxplots don’t look good as we only have two values per group. We
could just plot the individual points instead. We could use
<code>geom_point()</code> to make a scatterplot.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-37-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>The points are overlapping so we will make a jitter plot using
<code>geom_jitter()</code>. A <strong>jitter plot</strong> is similar to
a scatter plot. It adds a small amount of random variation to the
location of each point so they don’t overlap. If is also quite common to
combine jitter plots with other types of plot, for example, <a href="https://simplystatistics.org/2019/02/21/dynamite-plots-must-die/" class="external-link">jitter
with boxplot</a>.</p>
<div class="codewrapper sourceCode" id="cb58">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-38-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can colour the groups similar to before using
<code>colour =</code>.</p>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-39-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
<div class="section level1">
<h1 id="customising-the-plot">Customising the plot<a class="anchor" aria-label="anchor" href="#customising-the-plot"></a>
</h1>
<div class="section level2">
<h2 id="specifying-colours">Specifying colours<a class="anchor" aria-label="anchor" href="#specifying-colours"></a>
</h2>
<p>We might want to change the colours. To see what colour names are
available you can type <code>colours()</code>. There is also an <a href="https://www.nceas.ucsb.edu/~frazier/RSpatialGuides/colorPaletteCheatsheet.pdf" class="external-link">R
colours cheatsheet</a> that shows what the colours look like.</p>
<div class="codewrapper sourceCode" id="cb60">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mycolours</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"turquoise"</span>, <span class="st">"plum"</span>, <span class="st">"tomato"</span>, <span class="st">"violet"</span>, <span class="st">"steelblue"</span>, <span class="st">"chocolate"</span><span class="op">)</span></span></code></pre>
</div>
<p>Then we then add these colours to the plot using a <code>+</code> and
<code>scale_colour_manual(values = mycolours)</code>.</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">mycolours</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-41-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are built-in colour palettes that can be handy to use, where
the sets of colours are predefined. <code>scale_colour_brewer()</code>
is a popular one (there is also <code>scale_fill_brewer()</code>). You
can take a look at the help for <code>scale_colour_brewer()</code> to
see what palettes are available. The <a href="https://www.nceas.ucsb.edu/~frazier/RSpatialGuides/colorPaletteCheatsheet.pdf" class="external-link">R
colours cheatsheet</a> also shows what the colours of the palettes look
like. There’s one called “Dark2”, let’s have a look at that.</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">scale_colour_brewer</span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-42-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level2">
<h2 id="axis-labels-and-title">Axis labels and Title<a class="anchor" aria-label="anchor" href="#axis-labels-and-title"></a>
</h2>
<p>We can change the axis labels and add a title with
<code>labs()</code>. To change the x axis label we use
<code>labs(x = "New name")</code>. To change the y axis label we use
<code>labs(y = "New name")</code> or we can change them all at the same
time.</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-43-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level2">
<h2 id="themes">Themes<a class="anchor" aria-label="anchor" href="#themes"></a>
</h2>
<p>We can adjust the text on the x axis (the group labels) by turning
them 90 degrees so we can read the labels better. To do this we modify
the ggplot theme. Themes are the non-data parts of the plot.</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-44-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We can remove the grey background and grid lines.</p>
<p>There are also a lot of built-in themes. Let’s have a look at a
couple of the more widely used themes. The default ggplot theme is
<code>theme_grey().</code></p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-45-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-46-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are many themes available, you can see some in the <a href="https://www.r-graph-gallery.com/192-ggplot-themes/" class="external-link">R graph
gallery</a>.</p>
<p>We can also modify parts of the theme individually. We can remove the
grey background and grid lines with the code below.</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-47-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
<div class="section level2">
<h2 id="order-of-categories">Order of categories<a class="anchor" aria-label="anchor" href="#order-of-categories"></a>
</h2>
<p>The groups have been plotted in alphabetical order on the x axis and
in the legend (that is the default order), however, we may want to
change the order. We may prefer to plot the groups in order of stage,
for example, basal virgin, basal pregnant, basal lactate, luminal
virgin, luminal pregnant, luminal lactate.</p>
<p>First let’s make an object with the group order that we want.</p>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">group_order</span> <span class="op">&lt;-</span> <span class="fu">c</span><span class="op">(</span><span class="st">"bvirg"</span>, <span class="st">"bpreg"</span>, <span class="st">"blact"</span>, <span class="st">"lvirg"</span>, <span class="st">"lpreg"</span>, <span class="st">"llact"</span><span class="op">)</span></span></code></pre>
</div>
<p>Next we need to make a column with the groups into an R data type
called a <strong>factor</strong>. Factors in R are a special data type
used to specify categories, you can read more about them in the <a href="https://r4ds.had.co.nz/vectors.html#augmented-vectors" class="external-link">R for Data
Science book</a>. The names of the categories are called the factor
<strong>levels</strong>.</p>
<p>We’ll add another column called “Group_f” where we’ll make the Group
column into a factor and specify what order we want the levels of the
factor.</p>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes_counts</span> <span class="op">&lt;-</span> <span class="fu">mutate</span><span class="op">(</span><span class="va">mygenes_counts</span>, Group_f <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Group</span>, levels <span class="op">=</span> <span class="va">group_order</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Take a look at the data. As the table is quite wide we can use
<code>select()</code> to select just the columns we want to view.</p>
<div class="codewrapper sourceCode" id="cb70">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenes_counts</span> <span class="op">%&gt;%</span> <span class="fu">select</span><span class="op">(</span><span class="va">...1</span>, <span class="va">Group</span>, <span class="va">Group_f</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 84 × 3
   ...1               Group Group_f
   &lt;chr&gt;              &lt;chr&gt; &lt;fct&gt;
 1 ENSMUSG00000000381 lvirg lvirg
 2 ENSMUSG00000000381 lvirg lvirg
 3 ENSMUSG00000000381 lpreg lpreg
 4 ENSMUSG00000000381 lpreg lpreg
 5 ENSMUSG00000000381 llact llact
 6 ENSMUSG00000000381 llact llact
 7 ENSMUSG00000000381 bvirg bvirg
 8 ENSMUSG00000000381 bvirg bvirg
 9 ENSMUSG00000000381 bpreg bpreg
10 ENSMUSG00000000381 bpreg bpreg
# ℹ 74 more rows</code></pre>
</div>
<p>Notice that the Group column has <code>&lt;chr&gt;</code> under the
heading, that indicates is a character data type, while the Group_f
column has <code>&lt;fct&gt;</code> under the heading, indicating it is
a factor data type. The <code>str()</code> command that we saw
previously is useful to check the data types in objects.</p>
<div class="codewrapper sourceCode" id="cb72">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">mygenes_counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>tibble [84 × 9] (S3: tbl_df/tbl/data.frame)
 $ ...1               : chr [1:84] "ENSMUSG00000000381" "ENSMUSG00000000381" "ENSMUSG00000000381" "ENSMUSG00000000381" ...
 $ gene_symbol        : chr [1:84] "Wap" "Wap" "Wap" "Wap" ...
 $ Sample             : chr [1:84] "GSM1480291" "GSM1480292" "GSM1480293" "GSM1480294" ...
 $ Count              : num [1:84] 90.2 95.6 4140.3 8414.4 49204.9 ...
 $ characteristics    : chr [1:84] "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, 18.5 day pregnancy" "mammary gland, luminal cells, 18.5 day pregnancy" ...
 $ immunophenotype    : chr [1:84] "luminal cell population" "luminal cell population" "luminal cell population" "luminal cell population" ...
 $ developmental stage: chr [1:84] "virgin" "virgin" "18.5 day pregnancy" "18.5 day pregnancy" ...
 $ Group              : chr [1:84] "lvirg" "lvirg" "lpreg" "lpreg" ...
 $ Group_f            : Factor w/ 6 levels "bvirg","bpreg",..: 4 4 5 5 6 6 1 1 2 2 ...</code></pre>
</div>
<p><code>str()</code> shows us Group_f column is a Factor with 6 levels
(categories).</p>
<p>We can check the factor levels of a column as below.</p>
<div class="codewrapper sourceCode" id="cb74">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">levels</span><span class="op">(</span><span class="va">mygenes_counts</span><span class="op">$</span><span class="va">Group_f</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "bvirg" "bpreg" "blact" "lvirg" "lpreg" "llact"</code></pre>
</div>
<p>The levels are in the order that we want, so we can now change our
plot to use the “Group_f” column instead of Group column (change
<code>x =</code> and <code>colour =</code>).</p>
<div class="codewrapper sourceCode" id="cb76">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group_f</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group_f</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="../fig/introduction-rendered-unnamed-chunk-53-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We could do similar if we wanted to have the genes in the facets in a
different order. For example, we could add another column called
“gene_symbol_f” where we make the gene_symbol column into a factor,
specifying the order of the levels.</p>
<div id="exercise-3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-3" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li><p>Make a colourblind-friendly plot using the colourblind-friendly
palettes <a href="https://www.cookbook-r.com/Graphs/Colors_(ggplot2)/#a-colorblind-friendly-palette" class="external-link">here</a>.</p></li>
<li><p>Create a plot (any plot whatsoever) and share it with the class
by pasting the image in the Google Docs link provided in your workshop.
You plot should use the <code>subtitle</code> argument in the
<code>labs</code> function to add a unique identifier (e.g. a message
and your name or initials) which is displayed below the title.</p></li>
</ol>
<p>Tip: An easy way to copy your plot in RStudio is using the plot
pane’s export option and selecting “Copy to Clipboard…”. You can then
paste it into the provided Google document.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="saving-plots">Saving plots<a class="anchor" aria-label="anchor" href="#saving-plots"></a>
</h1>
<p>We can save plots interactively by clicking Export in the Plots
window and saving as e.g. “myplot.pdf”. Or we can output plots to pdf
using <code>pdf()</code> followed by <code>dev.off()</code>. We put our
plot code after the call to <code>pdf()</code> and before closing the
plot device with <code>dev.off()</code>.</p>
<p>Let’s save our last plot.</p>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pdf</span><span class="op">(</span><span class="st">"myplot.pdf"</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span>data <span class="op">=</span> <span class="va">mygenes_counts</span>, </span>
<span>       mapping <span class="op">=</span> <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group_f</span>, y <span class="op">=</span> <span class="fu">log2</span><span class="op">(</span><span class="va">Count</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span>, colour <span class="op">=</span> <span class="va">Group_f</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span> <span class="va">gene_symbol</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Cell type and stage"</span>, y <span class="op">=</span> <span class="st">"Count"</span>, title <span class="op">=</span> <span class="st">"Mammary gland RNA-seq data"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>panel.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.major <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dev.off</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div id="exercise-4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="exercise-4" class="callout-inner">
<h3 class="callout-title">Exercise</h3>
<div class="callout-content">
<ol style="list-style-type: decimal">
<li>Download the raw counts for this dataset from GREIN.
<ol style="list-style-type: lower-alpha">
<li>Make a boxplot. Do the samples look any different to the normalised
counts?</li>
<li>Make subplots for the same set of 8 genes. Do they look any
different to the normalised counts?</li>
</ol>
</li>
<li>Download the normalised counts for the GSE63310 dataset from GREIN.
Make boxplots colouring the samples using different columns in the
metadata file.</li>
</ol>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul>
<li>Tabular data can be loaded into R with the tidyverse functions
<code>read_csv()</code> and <code>read_tsv()</code>
</li>
<li>Tidyverse functions such as <code>pivot_longer()</code>,
<code>mutate()</code>, <code>filter()</code>, <code>select()</code>,
<code>full_join()</code> can be used to manipulate data</li>
<li>A ggplot has 3 components: data (dataset), mapping (columns to plot)
and geom (type of plot). Different types of plots include
<code>geom_point()</code>, <code>geom_jitter()</code>,
<code>geom_line()</code>, <code>geom_boxplot()</code>,
<code>geom_violin()</code>.</li>
<li>
<code>facet_wrap()</code> can be used to make subplots of the
data</li>
<li>The aesthetics of a ggplot can be modified, such as colouring by
different columns in the dataset, adding labels or changing the
background</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="further-reading">Further Reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h1>
<ul>
<li>
<a href="https://pmacdasci.github.io/r-intro-tidyverse/" class="external-link">A short
intro to R and tidyverse</a><br>
</li>
<li>
<a href="https://r-statistics.co/Top50-Ggplot2-Visualizations-MasterList-R-Code.html" class="external-link">Top
50 Ggplot Visualisations</a><br>
</li>
<li><a href="https://r4ds.had.co.nz/" class="external-link">R for Data Science</a></li>
</ul>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</div></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/edit/main/README.md" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/melbournebioinformatics/workbench-template-rmd/" class="external-link">Source</a></p>
				<p><a href="https://github.com/melbournebioinformatics/workbench-template-rmd/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:fixme@fixme.org">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.7" class="external-link">pegboard (0.7.7)</a>, and <a href="https://github.com/melbournebioinformatics/mb-varnish/tree/main" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "melbournebioinformatics.github.io/uom-varnish/instructor/aio.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "melbournebioinformatics.github.io/uom-varnish/instructor/aio.html",
  "identifier": "melbournebioinformatics.github.io/uom-varnish/instructor/aio.html",
  "dateCreated": "2025-01-29",
  "dateModified": "2025-02-11",
  "datePublished": "2025-02-11"
}

  </script><script>
		feather.replace();
	</script>
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

