<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en" data-bs-theme="auto"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to R for Biologists: Exploring the Data</title><meta name="viewport" content="width=device-width, initial-scale=1"><script src="../assets/themetoggle.js"></script><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../favicons/uom/apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicons/uom/favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicons/uom/favicon-16x16.png"><link rel="manifest" href="../favicons/uom/site.webmanifest"><link rel="mask-icon" href="../favicons/uom/safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" media="(prefers-color-scheme: light)" content="white"><meta name="theme-color" media="(prefers-color-scheme: dark)" content="black"></head><body>
    <header id="top" class="navbar navbar-expand-md top-nav uom"><svg xmlns="http://www.w3.org/2000/svg" class="d-none"><symbol id="check2" viewbox="0 0 16 16"><path d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0z"></path></symbol><symbol id="circle-half" viewbox="0 0 16 16"><path d="M8 15A7 7 0 1 0 8 1v14zm0 1A8 8 0 1 1 8 0a8 8 0 0 1 0 16z"></path></symbol><symbol id="moon-stars-fill" viewbox="0 0 16 16"><path d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"></path><path d="M10.794 3.148a.217.217 0 0 1 .412 0l.387 1.162c.173.518.579.924 1.097 1.097l1.162.387a.217.217 0 0 1 0 .412l-1.162.387a1.734 1.734 0 0 0-1.097 1.097l-.387 1.162a.217.217 0 0 1-.412 0l-.387-1.162A1.734 1.734 0 0 0 9.31 6.593l-1.162-.387a.217.217 0 0 1 0-.412l1.162-.387a1.734 1.734 0 0 0 1.097-1.097l.387-1.162zM13.863.099a.145.145 0 0 1 .274 0l.258.774c.115.346.386.617.732.732l.774.258a.145.145 0 0 1 0 .274l-.774.258a1.156 1.156 0 0 0-.732.732l-.258.774a.145.145 0 0 1-.274 0l-.258-.774a1.156 1.156 0 0 0-.732-.732l-.774-.258a.145.145 0 0 1 0-.274l.774-.258c.346-.115.617-.386.732-.732L13.863.1z"></path></symbol><symbol id="sun-fill" viewbox="0 0 16 16"><path d="M8 12a4 4 0 1 0 0-8 4 4 0 0 0 0 8zM8 0a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 0zm0 13a.5.5 0 0 1 .5.5v2a.5.5 0 0 1-1 0v-2A.5.5 0 0 1 8 13zm8-5a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2a.5.5 0 0 1 .5.5zM3 8a.5.5 0 0 1-.5.5h-2a.5.5 0 0 1 0-1h2A.5.5 0 0 1 3 8zm10.657-5.657a.5.5 0 0 1 0 .707l-1.414 1.415a.5.5 0 1 1-.707-.708l1.414-1.414a.5.5 0 0 1 .707 0zm-9.193 9.193a.5.5 0 0 1 0 .707L3.05 13.657a.5.5 0 0 1-.707-.707l1.414-1.414a.5.5 0 0 1 .707 0zm9.193 2.121a.5.5 0 0 1-.707 0l-1.414-1.414a.5.5 0 0 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .707zM4.464 4.465a.5.5 0 0 1-.707 0L2.343 3.05a.5.5 0 1 1 .707-.707l1.414 1.414a.5.5 0 0 1 0 .708z"></path></symbol></svg><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-8">
      <div class="large-logo">
        <img id="uom-logo" alt="Lesson Description" src="../assets/images/uom-logo.svg"><span class="badge text-bg-danger">
          <abbr title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.">
            <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="external-link alert-link">
              <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="border-radius: 5px"></i>
              Pre-Alpha
            </a>
            <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
          </abbr>
        </span>

      </div>
    </div>
    <div class="selector-container">
      <div id="theme-selector">
        <li class="nav-item dropdown" id="theme-button-list">
          <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
            <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="bd-theme-text"><li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                Light
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                Dark
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
            <li>
              <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                Auto
                <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
            </li>
          </ul></li>
      </div>

      <div class="dropdown" id="instructor-dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../explore-data.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl bottom-nav uom" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle Navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Lesson Description" src="../assets/images/uom-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to R for Biologists
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
        <i role="img" aria-label="Search the All In One page" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to R for Biologists
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
          </ul></li>
      </ul></div>
    <!--
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
      <input class="form-control me-2 searchbox" type="search" placeholder="" aria-label="">
        <button class="btn btn-outline-success tablet-search-button"  type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="Search the All In One page"></i>
        </button>
      </fieldset>
    </form>
    -->
    <a id="search-button" class="btn btn-primary" href="../instructor/aio.html" role="button" aria-label="Search the All In One page">Search the All In One page</a>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to R for Biologists
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 24%" class="percentage">
    24%
  </div>
  <div class="progress uom">
    <div class="progress-bar uom" role="progressbar" style="width: 24%" aria-valuenow="24" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
<div id="sidebar-col" class="col-lg-4">
  <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle" data-collapse="Collapse " data-episodes="Episodes ">
          <i class="search-icon" data-feather="x" role="img"></i>
        </button>
        <div class="sidebar-inner">
          <div class="row mobile-row" id="theme-row-mobile">
            <div class="col" id="theme-selector">
              <li class="nav-item dropdown" id="theme-button-list">
                <button class="btn btn-link nav-link px-0 px-lg-2 dropdown-toggle d-flex align-items-center" id="bd-theme" type="button" aria-expanded="false" data-bs-toggle="dropdown" data-bs-display="static" aria-label="Toggle theme (auto)">
                  <svg class="bi my-1 theme-icon-active"><use href="#circle-half"></use></svg><span class="d-lg-none ms-1" id="bd-theme-text">Toggle Theme</span>
                </button>
                <ul class="dropdown-menu dropdown-menu-right" aria-labelledby="bd-theme-text"><li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="light" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#sun-fill"></use></svg>
                      Light
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center" data-bs-theme-value="dark" aria-pressed="false">
                      <svg class="bi me-2 theme-icon"><use href="#moon-stars-fill"></use></svg>
                      Dark
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                  <li>
                    <button type="button" class="btn dropdown-item d-flex align-items-center active" data-bs-theme-value="auto" aria-pressed="true">
                      <svg class="bi me-2 theme-icon"><use href="#circle-half"></use></svg>
                      Auto
                      <svg class="bi ms-auto d-none"><use href="#check2"></use></svg></button>
                  </li>
                </ul></li>
            </div>
          </div>
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../explore-data.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="introduction.html">1. Introduction to R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        2. Exploring the Data
        </span>
      </button>
    </div><!--/div.accordion-header-->

    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#getting-started-with-the-data">Getting started with the data</a></li>
<li><a href="#rna-seq-dataset-from-fu-et-al.">RNA-seq dataset from Fu et al.</a></li>
<li><a href="#tidyverse">Tidyverse</a></li>
<li><a href="#data-files">Data files</a></li>
<li><a href="#loading-the-data">Loading the data</a></li>
<li><a href="#getting-to-know-the-data">Getting to know the data</a></li>
<li><a href="#dimensions-of-the-data">Dimensions of the data</a></li>
<li><a href="#column-and-row-names-of-the-data">Column and row names of the data</a></li>
<li><a href="#subsetting">Subsetting</a></li>
<li><a href="#renaming-column-names">Renaming column names</a></li>
<li><a href="#structure-and-summary">Structure and Summary</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="format-data.html">3. Formatting the Data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="ggplot2-intro.html">4. Introduction to ggplot2</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="ggplot2-customisation.html">5. Extra ggplot2 Customisation</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="wrap-up.html">6. Wrapping Up</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush lesson-resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="lesson-resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      <hr><li><a class="dropdown-item" href="reference.html">Reference</a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width lesson-resources"><a href="../instructor/aio.html">See all in one page</a>


            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/format-data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to R
        </a>
        <a class="chapter-link float-end" href="../instructor/format-data.html" rel="next">
          Next: Formatting the Data...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Exploring the Data</h1>
        <p>Last updated on 2025-02-18 |

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/explore-data.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <p>Estimated time: <i aria-hidden="true" data-feather="clock"></i> 40 minutes</p>

        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right" data-expand="Expand All Solutions " data-collapse="Collapse All Solutions "> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>How can I load data from CSV or TSV files into R?</li>
<li>What are some functions in R that can be used to examine the
data?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Load in our RNA-seq data files into R</li>
<li>Try using some functions to explore data frames (tables) in R</li>
<li>Learn how to subset parts of a data frame</li>
</ul></div>
</div>
</div>
</div>
</div>
<section><h2 class="section-heading" id="getting-started-with-the-data">Getting started with the data<a class="anchor" aria-label="anchor" href="#getting-started-with-the-data"></a></h2>
<hr class="half-width"><p>In this tutorial, we will learn some R through creating plots to
visualise data from an RNA-seq experiment.</p>
<p>The <a href="https://www.nature.com/articles/s41598-019-43935-8" class="external-link">GREIN
platform</a> (GEO RNA-seq Experiments Interactive Navigator) provides
&gt;6,500 published datasets from GEO that have been uniformly
processed. It is available at <a href="http://www.ilincs.org/apps/grein/" class="external-link uri">http://www.ilincs.org/apps/grein/</a>. You can search for a
dataset of interest using the GEO code. GREIN provide QC metrics for the
RNA-seq datasets and both raw and normalized counts. We will use the
normalized counts here. These are the counts of reads for each gene for
each sample normalized for differences in sequencing depth and
composition bias. Generally, the higher the number of counts the more
the gene is expressed.</p>
</section><section><h2 class="section-heading" id="rna-seq-dataset-from-fu-et-al-">RNA-seq dataset from Fu et al.<a class="anchor" aria-label="anchor" href="#rna-seq-dataset-from-fu-et-al-"></a></h2>
<hr class="half-width"><p>Here we will create some plots using RNA-seq data from the paper by
<a href="https://www.ncbi.nlm.nih.gov/pubmed/25730472" class="external-link">Fu et
al. 2015</a> (GEO accession number GSE60450). Mice, like all mammals,
have mammary glands which produce milk to nourish their young. The
authors of this study were interested in examining how the mammary
epithelium (which line the mammary glands) expands and develops during
pregnancy and lactation.</p>
<p>This study examined expression in basal and luminal cells from mice
at different stages (virgin, pregnant and lactating). Basal cells are
mammary stem cells and luminal cells secrete milk. There are 2 samples
per group and 6 groups, so 12 samples in total.</p>
<figure><img src="../fig/mouse_exp.png" class="figure mx-auto d-block"></figure><hr></section><section><h2 class="section-heading" id="tidyverse">Tidyverse<a class="anchor" aria-label="anchor" href="#tidyverse"></a></h2>
<hr class="half-width"><figure><img src="../fig/tidyverse.png" style="width:100.0%" class="figure mx-auto d-block"></figure><p>The <code>tidyverse</code> package that we installed previously is a
collection of R packages that includes the extremely widely used
<strong><code>ggplot2</code></strong> package.</p>
<p>The <a href="https://www.tidyverse.org/" class="external-link">tidyverse</a> makes data
science faster, easier and more fun.</p>
<figure><img src="../fig/r_tidyverse.png" style="width:100.0%" class="figure mx-auto d-block"></figure><p>Tidyverse is built on the principle of organizing data in a <a href="https://www.openscapes.org/blog/2020/10/12/tidy-data/" class="external-link">tidy
format</a>.</p>
<hr></section><section><h2 class="section-heading" id="data-files">Data files<a class="anchor" aria-label="anchor" href="#data-files"></a></h2>
<hr class="half-width"><div id="your-turn-2.1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.1" class="callout-inner">
<h3 class="callout-title">Your turn 2.1</h3>
<div class="callout-content">
<p>If you haven’t already downloaded the data.zip file for this
workshop, you can click <a href="../data/data.zip">here</a> to download it
now.</p>
<p>Unzip the file and store the extracted <code>data</code> folder in
your working directory.</p>
</div>
</div>
</div>
<div id="make-sure-the-data-is-in-the-correct-directory" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="make-sure-the-data-is-in-the-correct-directory" class="callout-inner">
<h3 class="callout-title">Make sure the data is in the correct directory</h3>
<div class="callout-content">
<p>Inside your current working directory directory
(e.g. <code>intro_r_course</code>), there should be a directory called
<code>data</code> and inside that directory should be two CSV files.</p>
<figure><img src="../fig/working_dir_files.png" style="width:80.0%" class="figure mx-auto d-block"></figure><p>In the next section, you will load these files into R using the file
path (i.e. where they are located in the filesystem), so the code in
this tutorial expects the files to be in the above location. If you have
the files in a different location, you also have the option of changing
the commands in the following sections to match where the files are
located on your computer.</p>
</div>
</div>
</div>
</section><section><h2 class="section-heading" id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a></h2>
<hr class="half-width"><p>We use <code>library()</code> to load in the packages that we need.
As described in the cooking analogy in the first screenshot,
<code>install.packages()</code> is like buying a saucepan,
<code>library()</code> is taking it out of the cupboard to use it.</p>
<div id="your-turn-2.2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.2" class="callout-inner">
<h3 class="callout-title">Your turn 2.2</h3>
<div class="callout-content">
<p>Load in the tidyverse package using the <code>library()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">tidyverse</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>The files we will use are CSV comma-separated, so we will use the
<code>read_csv()</code> function from the tidyverse. There is also a
<code>read_tsv()</code> function for tab-separated values.</p>
<p>We will use the counts file called
<code>GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv</code> that’s
in a folder called <code>data</code> i.e. the path to the file should be
<code>data/GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv</code>.</p>
<p>We can read the counts file into R with the command below. We’ll
store the contents of the counts file in an <strong>object</strong>
called <code>counts</code>. This stores the file contents in R’s memory
making it easier to use.</p>
<div id="your-turn-2.3" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.3" class="callout-inner">
<h3 class="callout-title">Your turn 2.3</h3>
<div class="callout-content">
<p>Load the count data into R. We will store the contents of the counts
file in an <strong>object</strong> called <code>counts</code>. Note that
we need to put quotes (““) around file paths.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in counts file</span></span>
<span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/GSE60450_GeneLevel_Normalized(CPM.and.TMM)_data.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>New names:
Rows: 23735 Columns: 14
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(2): ...1, gene_symbol dbl (12): GSM1480291, GSM1480292, GSM1480293,
GSM1480294, GSM1480295, GSM148...
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `` -&gt; `...1`</code></pre>
</div>
<p>Note: In R, structured tables like these are called <strong>data
frames</strong>.</p>
</div>
</div>
</div>
<div id="tab-completion-for-file-paths" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tab-completion-for-file-paths" class="callout-inner">
<h3 class="callout-title">Tab completion for file paths</h3>
<div class="callout-content">
<p>We mentioned tab completion in the previous section, but tab
completion can also complete file paths. This means you don’t have to
type out the long filenames in the codeblock above, but instead, begin
to type out a few characters, then press <kbd>tab</kbd> and see the
autocompletion options.</p>
<figure><img src="../fig/file_path_autocompletion.png" class="figure mx-auto d-block"></figure></div>
</div>
</div>
<p>No need to be overwhelmed by the outputs! It contains information
regarding “column specification” (telling us that there is a missing
column name in the header and it has been filled with the name “…1”,
which is how read_csv handles missing column names by default). We will
fix this later. It also tells us what data types <code>read_csv</code>
is detecting in each column. Columns with text characters have been
detected (<code>col_character</code>) and also columns with numbers
(<code>col_double</code>). We won’t get into the details of R data types
in this tutorial but they are important to know when you get more
proficient in R. You can read more about them in the <a href="https://r4ds.had.co.nz/vectors.html#important-types-of-atomic-vector" class="external-link">R
for Data Science book</a>.</p>
<div id="your-turn-2.4" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.4" class="callout-inner">
<h3 class="callout-title">Your turn 2.4</h3>
<div class="callout-content">
<p>Load the sample information data into R. We will store the contents
of this file in an object called <code>sampleinfo</code>.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in metadata</span></span>
<span><span class="va">sampleinfo</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/GSE60450_filtered_metadata.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>New names:
Rows: 12 Columns: 4
── Column specification
──────────────────────────────────────────────────────── Delimiter: "," chr
(4): ...1, characteristics, immunophenotype, developmental stage
ℹ Use `spec()` to retrieve the full column specification for this data. ℹ
Specify the column types or set `show_col_types = FALSE` to quiet this message.
• `` -&gt; `...1`</code></pre>
</div>
</div>
</div>
</div>
<p>It is very common when looking at biological data that you have two
types of data. One is the actual data (in this case, our
<code>counts</code> object, which has the expression values of different
genes in each sample). The other is <strong>metadata</strong>
i.e. information about our samples (in this case, our
<code>sampleinfo</code> object includes information about whether
samples are from basal or luminal cells and whether the cells were from
mice which are virgin/pregnant/lactating, etc.)</p>
<p><strong>What data have we imported into R?</strong><br>
To summarise, we have imported two data frames (i.e. tables) into R:</p>
<ul><li>
<code>counts</code> object is our gene expression data</li>
<li>
<code>sampleinfo</code> object is our sample metadata</li>
</ul><div id="your-turn-2.5" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.5" class="callout-inner">
<h3 class="callout-title">Your turn 2.5</h3>
<div class="callout-content">
<p>Let’s get used to making some mistakes in R, so we know what errors
look like and how to handle them.</p>
<ol style="list-style-type: decimal"><li><p>Test what happens if you type
<code>Library(tidyverse)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type
<code>library(tidyverse</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_tsv("data/GSE60450_filtered_metadata.csv")</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_csv("data/GSE60450_filtered_metadata.csv)</code><br>
What is wrong and how would you fix it?</p></li>
<li><p>Test what happens if you type<br><code>read_csv("GSE60450_filtered_metadata.csv")</code><br>
What is wrong and how would you fix it?</p></li>
</ol><p>Don’t forget you can press <kbd>ESC</kbd> to escape the current
command and start a new prompt.</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="getting-to-know-the-data">Getting to know the data<a class="anchor" aria-label="anchor" href="#getting-to-know-the-data"></a></h2>
<hr class="half-width"><p>When assigning a value to an object, R does not print the value. We
do not see what is in <code>counts</code> or <code>sampleinfo</code>.
But there are ways we can look at the data.</p>
<div id="your-turn-2.6" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.6" class="callout-inner">
<h3 class="callout-title">Your turn 2.6</h3>
<div class="callout-content">
<p>Click on the <code>sampleinfo</code> object in your global
environment panel on the right-hand-side of RStudio. This will open a
new tab.</p>
<p>This is the equivalent of using the <code>View()</code> function.
e.g.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">View</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-2.7" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.7" class="callout-inner">
<h3 class="callout-title">Your turn 2.7</h3>
<div class="callout-content">
<p>Type the name of the object and this will print the first few lines
and some information, such as number of rows. Note that this is similar
to how we looked at the value of objects we assigned in the previous
section.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 12 × 4
   ...1       characteristics              immunophenotype `developmental stage`
   &lt;chr&gt;      &lt;chr&gt;                        &lt;chr&gt;           &lt;chr&gt;
 1 GSM1480291 mammary gland, luminal cell… luminal cell p… virgin
 2 GSM1480292 mammary gland, luminal cell… luminal cell p… virgin
 3 GSM1480293 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 4 GSM1480294 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 5 GSM1480295 mammary gland, luminal cell… luminal cell p… 2 day lactation
 6 GSM1480296 mammary gland, luminal cell… luminal cell p… 2 day lactation
 7 GSM1480297 mammary gland, basal cells,… basal cell pop… virgin
 8 GSM1480298 mammary gland, basal cells,… basal cell pop… virgin
 9 GSM1480299 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
10 GSM1480300 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
11 GSM1480301 mammary gland, basal cells,… basal cell pop… 2 day lactation
12 GSM1480302 mammary gland, basal cells,… basal cell pop… 2 day lactation      </code></pre>
</div>
</div>
</div>
</div>
<p>We can also take a look the first few lines with <code>head()</code>.
This shows us the first 6 lines.</p>
<div id="your-turn-2.8" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.8" class="callout-inner">
<h3 class="callout-title">Your turn 2.8</h3>
<div class="callout-content">
<p>Use <code>head()</code> to look at the first few lines of
<code>counts</code>.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 14
  ...1        gene_symbol GSM1480291 GSM1480292 GSM1480293 GSM1480294 GSM1480295
  &lt;chr&gt;       &lt;chr&gt;            &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 ENSMUSG000… Gnai3          243.      256.         240.       217.       84.7
2 ENSMUSG000… Pbsn             0         0            0          0         0
3 ENSMUSG000… Cdc45           11.2      13.8         11.6        4.27      8.35
4 ENSMUSG000… H19              6.31      8.53         7.09      11.0       0.194
5 ENSMUSG000… Scml2            2.19      4.66         2.80       2.50      1.24
6 ENSMUSG000… Apoh             0.224     0.0840       0          0         0
# ℹ 7 more variables: GSM1480296 &lt;dbl&gt;, GSM1480297 &lt;dbl&gt;, GSM1480298 &lt;dbl&gt;,
#   GSM1480299 &lt;dbl&gt;, GSM1480300 &lt;dbl&gt;, GSM1480301 &lt;dbl&gt;, GSM1480302 &lt;dbl&gt;</code></pre>
</div>
</div>
</div>
</div>
<p>We can also look at the last few lines with <code>tail()</code>. This
shows us the last 6 lines. This can be useful to check the bottom of the
file, that it looks ok.</p>
<div id="your-turn-2.6-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.6-1" class="callout-inner">
<h3 class="callout-title">Your turn 2.6</h3>
<div class="callout-content">
<p>Use <code>tail()</code> to look at the last few lines of
<code>counts</code>.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">tail</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 14
  ...1        gene_symbol GSM1480291 GSM1480292 GSM1480293 GSM1480294 GSM1480295
  &lt;chr&gt;       &lt;chr&gt;            &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;
1 ENSMUSG000… Pcdha7           0.134     0          0            0          0
2 ENSMUSG000… Gm34240          0         0          0            0          0
3 ENSMUSG000… Pcdhga3          0.582     2.27       0.334        1.10       0
4 ENSMUSG000… Gm20750          0         0.0840     0.0417       0          0
5 ENSMUSG000… Rhbg             5.28      4.96       6.26         3.98       1.01
6 ENSMUSG000… Mat2a          212.      225.        97.9         70.1       22.0
# ℹ 7 more variables: GSM1480296 &lt;dbl&gt;, GSM1480297 &lt;dbl&gt;, GSM1480298 &lt;dbl&gt;,
#   GSM1480299 &lt;dbl&gt;, GSM1480300 &lt;dbl&gt;, GSM1480301 &lt;dbl&gt;, GSM1480302 &lt;dbl&gt;</code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-2.7-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="your-turn-2.7-1" class="callout-inner">
<h3 class="callout-title">Your turn 2.7</h3>
<div class="callout-content">
<p>What are the cell types of the first 6 samples in the metadata for
the Fu et al. 2015 experiment?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1"> Show me the solution </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 4
  ...1       characteristics               immunophenotype `developmental stage`
  &lt;chr&gt;      &lt;chr&gt;                         &lt;chr&gt;           &lt;chr&gt;
1 GSM1480291 mammary gland, luminal cells… luminal cell p… virgin
2 GSM1480292 mammary gland, luminal cells… luminal cell p… virgin
3 GSM1480293 mammary gland, luminal cells… luminal cell p… 18.5 day pregnancy
4 GSM1480294 mammary gland, luminal cells… luminal cell p… 18.5 day pregnancy
5 GSM1480295 mammary gland, luminal cells… luminal cell p… 2 day lactation
6 GSM1480296 mammary gland, luminal cells… luminal cell p… 2 day lactation      </code></pre>
</div>
<p>The first 6 samples are luminal cells.</p>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="dimensions-of-the-data">Dimensions of the data<a class="anchor" aria-label="anchor" href="#dimensions-of-the-data"></a></h2>
<hr class="half-width"><p>You can print the number of rows and columns using the function
<code>dim()</code>.</p>
<p>For example:</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 12  4</code></pre>
</div>
<p><code>sampleinfo</code> has 12 rows, corresponding to our 12 samples,
and 4 columns, corresponding to different features about the
samples.</p>
<div id="tip-always-verify-your-data-size" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="tip-always-verify-your-data-size" class="callout-inner">
<h3 class="callout-title">Tip: Always Verify Your Data Size</h3>
<div class="callout-content">
<p>Double-check that your data has the expected number of rows and
columns. It’s easy to read the wrong file or encounter corrupted
downloads. Catching these issues early will save you a lot of trouble
later!</p>
</div>
</div>
</div>
<div id="your-turn-2.8-1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="your-turn-2.8-1" class="callout-inner">
<h3 class="callout-title">Your turn 2.8</h3>
<div class="callout-content">
<p>Check how many rows and columns are in <code>counts</code>. Do these
numbers match what we expected?</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2"> Show me the solution </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dim</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 23735    14</code></pre>
</div>
<p>We know that there are 12 samples in our data (see the diagram
above), and we don’t know how many genes were measured.</p>
<p><strong>Rows:</strong> 23735. This means there are 23735 genes, which
could be correct (we didn’t know how many genes were measured).</p>
<p><strong>Columns:</strong> 14. We would expect to have 12 columns
corresponding to our 12 samples, but instead we have 14. Why is this?
Have a look back at when we visualised the <code>counts</code> data,
there are two extra columns in the data corresponding to gene IDs and
gene names.</p>
</div>
</div>
</div>
</div>
<p>In the Environment Tab in the top right panel in RStudio we can also
see the number of rows and columns in the objects we have in our
session.</p>
<hr></section><section><h2 class="section-heading" id="column-and-row-names-of-the-data">Column and row names of the data<a class="anchor" aria-label="anchor" href="#column-and-row-names-of-the-data"></a></h2>
<hr class="half-width"><div id="your-turn-2.9" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.9" class="callout-inner">
<h3 class="callout-title">Your turn 2.9</h3>
<div class="callout-content">
<p>Check the column and row names used in in <code>sampleinfo</code></p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="subsetting">Subsetting<a class="anchor" aria-label="anchor" href="#subsetting"></a></h2>
<hr class="half-width"><p>Subsetting is very useful tool in R which allows you to extract parts
of the data you want to analyse. There are multiple ways to subset data
and here we’ll only cover a few.</p>
<p>We can use the <code>$</code> operator to access individual columns
by name.</p>
<div id="your-turn-2.10" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.10" class="callout-inner">
<h3 class="callout-title">Your turn 2.10</h3>
<div class="callout-content">
<p>Extract the ‘immunophenotype’ column of the metadata.</p>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">$</span><span class="va">immunophenotype</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>We can also use square brackets <code>[ ]</code> to access the rows
and columns of a matrix or a table.</p>
<p>For example, we can extract the first row ‘1’ of the data, using the
number on the <em>left-hand-side</em> of the comma.</p>
<div id="your-turn-2.11" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.11" class="callout-inner">
<h3 class="callout-title">Your turn 2.11</h3>
<div class="callout-content">
<p>Extract the first row using square brackets.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fl">1</span>,<span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Here we extract the second column ‘2’ of the data, as indicated on
the <em>right-hand-side</em> of the comma.</p>
<div id="your-turn-2.12" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.12" class="callout-inner">
<h3 class="callout-title">Your turn 2.12</h3>
<div class="callout-content">
<p>Extract the second column using square brackets.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="common-r-error-missing-comma-in-subsetting" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="common-r-error-missing-comma-in-subsetting" class="callout-inner">
<h3 class="callout-title">Common R Error: Missing Comma in Subsetting</h3>
<div class="callout-content">
<p>When subsetting data frames or matrices (which have <em>two</em>
dimensions, i.e. rows and columns), always use a comma to separate rows
and columns, like <code>sampleinfo[1, ]</code>. Forgetting the comma, as
in <code>sampleinfo[1]</code>, will cause an error. Try it yourself and
see what happens!</p>
</div>
</div>
</div>
<p>You can use a combination of number of row and column to extract one
element in the matrix.</p>
<div id="your-turn-2.13" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.13" class="callout-inner">
<h3 class="callout-title">Your turn 2.13</h3>
<div class="callout-content">
<p>Extract the element in the first row and second column.</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="your-turn-2.14" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.14" class="callout-inner">
<h3 class="callout-title">Your turn 2.14</h3>
<div class="callout-content">
<p>We can also subset using a range of numbers. For example, if we
wanted the first three rows of <code>sampleinfo</code></p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="op">]</span></span></code></pre>
</div>
<p>Or if we wanted the 2nd, 4th, and 5th row:</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span><span class="op">[</span><span class="fu">c</span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">5</span><span class="op">)</span>,<span class="op">]</span></span></code></pre>
</div>
</div>
</div>
</div>
<div id="the-c-function" class="callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="the-c-function" class="callout-inner">
<h3 class="callout-title">The c() function</h3>
<div class="callout-content">
<p>We use the <code>c()</code> function extremely often in R when we
have multiple items that we are <em>combining</em> (‘c’ stands for
concatenating). We will see it again in this tutorial.</p>
</div>
</div>
</div>
<hr></section><section><h2 class="section-heading" id="renaming-column-names">Renaming column names<a class="anchor" aria-label="anchor" href="#renaming-column-names"></a></h2>
<hr class="half-width"><p>In the previous section, when we loaded in the data from the csv
file, we noticed that the first column had a missing column name and by
default, read_csv function assigned a name of “...1” to it. Let’s change
this column to something more descriptive now. We can do this by
combining a few things we’ve just learnt.</p>
<div id="your-turn-2.15" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.15" class="callout-inner">
<h3 class="callout-title">Your turn 2.15</h3>
<div class="callout-content">
<p>First, we use the <code>colnames()</code> function to obtain the
column names of sampleinfo. Then we use square brackets to subset the
first value of the column names (<code>[1]</code>). Last, we use the
assignment operator (<code>&lt;-</code>) to set the new value of the
first column name to “sample_id”.</p>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"sample_id"</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Let’s check if this has been changed correctly.</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sampleinfo</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 12 × 4
   sample_id  characteristics              immunophenotype `developmental stage`
   &lt;chr&gt;      &lt;chr&gt;                        &lt;chr&gt;           &lt;chr&gt;
 1 GSM1480291 mammary gland, luminal cell… luminal cell p… virgin
 2 GSM1480292 mammary gland, luminal cell… luminal cell p… virgin
 3 GSM1480293 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 4 GSM1480294 mammary gland, luminal cell… luminal cell p… 18.5 day pregnancy
 5 GSM1480295 mammary gland, luminal cell… luminal cell p… 2 day lactation
 6 GSM1480296 mammary gland, luminal cell… luminal cell p… 2 day lactation
 7 GSM1480297 mammary gland, basal cells,… basal cell pop… virgin
 8 GSM1480298 mammary gland, basal cells,… basal cell pop… virgin
 9 GSM1480299 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
10 GSM1480300 mammary gland, basal cells,… basal cell pop… 18.5 day pregnancy
11 GSM1480301 mammary gland, basal cells,… basal cell pop… 2 day lactation
12 GSM1480302 mammary gland, basal cells,… basal cell pop… 2 day lactation      </code></pre>
</div>
<p>The first column is now named “sample_id”.</p>
<p>We can also do the same to the counts data. This time, we rename the
first column name from “...1” to “gene_id”.</p>
<div id="your-turn-2.16" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="your-turn-2.16" class="callout-inner">
<h3 class="callout-title">Your turn 2.16</h3>
<div class="callout-content">
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"gene_id"</span></span></code></pre>
</div>
</div>
</div>
</div>
<p>Note: there are multiple ways to rename columns. We’ve covered one
way here, but another way is using the <code>rename()</code> function.
When programming, you’ll often find many ways to do the same thing.
Often there is one obvious method depending on the context you’re
in.</p>
<hr></section><section><h2 class="section-heading" id="structure-and-summary">Structure and Summary<a class="anchor" aria-label="anchor" href="#structure-and-summary"></a></h2>
<hr class="half-width"><p>Other useful commands for checking data are <code>str()</code> and
<code>summary()</code>.</p>
<p><code>str()</code> shows us the structure of our data. It shows us
what columns there are, the first few entries, and what data type they
are e.g. character or numbers (double or integer).</p>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">str</span><span class="op">(</span><span class="va">sampleinfo</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>spc_tbl_ [12 × 4] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ sample_id          : chr [1:12] "GSM1480291" "GSM1480292" "GSM1480293" "GSM1480294" ...
 $ characteristics    : chr [1:12] "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, virgin" "mammary gland, luminal cells, 18.5 day pregnancy" "mammary gland, luminal cells, 18.5 day pregnancy" ...
 $ immunophenotype    : chr [1:12] "luminal cell population" "luminal cell population" "luminal cell population" "luminal cell population" ...
 $ developmental stage: chr [1:12] "virgin" "virgin" "18.5 day pregnancy" "18.5 day pregnancy" ...
 - attr(*, "spec")=
  .. cols(
  ..   ...1 = col_character(),
  ..   characteristics = col_character(),
  ..   immunophenotype = col_character(),
  ..   `developmental stage` = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
<p><code>summary()</code> generates summary statistics of our data. For
numeric columns (columns of type double or integer) it outputs
statistics such as the min, max, mean and median. We will demonstrate
this with the counts file as it contains numeric data. For character
columns it shows us the length (how many rows).</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">summary</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>   gene_id          gene_symbol          GSM1480291          GSM1480292
 Length:23735       Length:23735       Min.   :    0.000   Min.   :    0.000
 Class :character   Class :character   1st Qu.:    0.000   1st Qu.:    0.000
 Mode  :character   Mode  :character   Median :    1.745   Median :    1.891
                                       Mean   :   42.132   Mean   :   42.132
                                       3rd Qu.:   29.840   3rd Qu.:   29.604
                                       Max.   :12525.066   Max.   :12416.211
   GSM1480293         GSM1480294         GSM1480295          GSM1480296
 Min.   :    0.00   Min.   :    0.00   Min.   :     0.00   Min.   :     0.00
 1st Qu.:    0.00   1st Qu.:    0.00   1st Qu.:     0.00   1st Qu.:     0.00
 Median :    0.92   Median :    0.89   Median :     0.58   Median :     0.54
 Mean   :   42.13   Mean   :   42.13   Mean   :    42.13   Mean   :    42.13
 3rd Qu.:   21.91   3rd Qu.:   19.92   3rd Qu.:    12.27   3rd Qu.:    12.28
 Max.   :49191.15   Max.   :55692.09   Max.   :111850.87   Max.   :108726.08
   GSM1480297          GSM1480298          GSM1480299
 Min.   :    0.000   Min.   :    0.000   Min.   :    0.000
 1st Qu.:    0.000   1st Qu.:    0.000   1st Qu.:    0.000
 Median :    2.158   Median :    2.254   Median :    1.854
 Mean   :   42.132   Mean   :   42.132   Mean   :   42.132
 3rd Qu.:   27.414   3rd Qu.:   26.450   3rd Qu.:   24.860
 Max.   :10489.311   Max.   :10662.486   Max.   :15194.048
   GSM1480300          GSM1480301          GSM1480302
 Min.   :    0.000   Min.   :    0.000   Min.   :    0.000
 1st Qu.:    0.000   1st Qu.:    0.000   1st Qu.:    0.000
 Median :    1.816   Median :    1.629   Median :    1.749
 Mean   :   42.132   Mean   :   42.132   Mean   :   42.132
 3rd Qu.:   23.443   3rd Qu.:   23.443   3rd Qu.:   24.818
 Max.   :17434.935   Max.   :19152.728   Max.   :15997.193  </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>The <code>read_csv()</code> and <code>read_tsv()</code> functions
can be used to load in CSV and TSV files in R</li>
<li>The <code>head()</code> and <code>tail()</code> functions can print
the first and last parts of an object and the <code>dim()</code>
function prints the dimensions of an object</li>
<li>Subsetting can be done with the <code>$</code> operator using column
names or using square brackets <code>[ ]</code>
</li>
<li>The <code>str()</code> and <code>summary()</code> functions are
useful functions to get an overview or summary of the data</li>
</ul></div>
</div>
</div>
<!--
Place links that you need to refer to multiple times across pages here. Delete
any links that you are not going to use.
 -->
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/introduction.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/format-data.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/introduction.html" rel="prev">
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>
          Previous: Introduction to R
        </a>
        <a class="chapter-link float-end" href="../instructor/format-data.html" rel="next">
          Next: Formatting the Data...
          <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i>
        </a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
        <p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/melbournebioinformatics/intro-to-R/edit/main/episodes/explore-data.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/melbournebioinformatics/intro-to-R/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/melbournebioinformatics/intro-to-R/" class="external-link">Source</a></p>
				<p><a href="https://github.com/melbournebioinformatics/intro-to-R/blob/main/CITATION.cff" class="external-link">Cite</a> | <a href="mailto:jchung@unimelb.edu.au">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p>Template licensed under <a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">CC-BY 4.0</a> by <a href="https://carpentries.org/" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.11" class="external-link">sandpaper (0.16.11)</a>, <a href="https://github.com/carpentries/pegboard/tree/0.7.7" class="external-link">pegboard (0.7.7)</a>, and <a href="https://github.com/melbournebioinformatics/uom-varnish/tree/main" class="external-link">varnish (1.0.5)</a></p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
      <i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back To Top"></i><br><!-- <span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top --><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "melbournebioinformatics.github.io/uom-varnish/instructor/explore-data.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "R, programming, plotting, ggplot, tidyverse, data",
  "name": "Exploring the Data",
  "creativeWorkStatus": "active",
  "url": "melbournebioinformatics.github.io/uom-varnish/instructor/explore-data.html",
  "identifier": "melbournebioinformatics.github.io/uom-varnish/instructor/explore-data.html",
  "dateCreated": "2025-02-18",
  "dateModified": "2025-02-18",
  "datePublished": "2025-02-18"
}

  </script><script>
		feather.replace();
	</script></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

